% \documentclass{article}
% \usepackage{fullpage}
% \input{review-response-commands}
% 
% \begin{document}


%%%%%%
%%
%%  Don't reorder the reviewer points; that'll mess up the automatic referencing!
%%
%%%%%

\begin{minipage}[b]{2.5in}
  Resubmission Cover Letter \\
  {\it Molecular Ecology}
\end{minipage}
\hfill
\begin{minipage}[b]{2.5in}
    Lewis-Sigler institute for integrative genomics\\
    Princeton University\\
    Yaniv's address\\
    \emph{and} Molecular and Computational Biology\\
    University of Southern California \\
  \today
\end{minipage}
 
\vskip 2em
 
\noindent
{\bf To the Editor(s) -- }
 
\vskip 1em


\yb{We are pleased to submit a revision of our Manuscript, `Beyond clines: lineages and haplotype blocks in hybrid zones' (MEC-16-0266) for the special issue 'Genomics of Hybridization'. We believe that this version is much improved by changes we have made to address reviewer and AE comments. Below we provide a point-by-point response to reviewer and AE comments, describing how we addressed concerns raised by our readers. The major themes of our changes are (\emph{1}) Attempting to make our mathematical derivations more accessible for Molecular Ecologyâ€™s wide audience  (\emph{2}) Additional justification for using a genetic model of under dominance to model hybrid incompatibilities, in which we make heavy reference to the historical theory justifying this approach, and (\emph{3}) Additional simulations to show that our results are not driven by boundaries imposed in a defined range, and (\emph{4})  A streamlined introduction that makes clear our major goals and provides a stronger structure for our manuscript. }

\begin{quote}
Reviewer 1 would like to see development of the finding that ancestral blocks are longer around a selected locus into a practical test.  I think that this would be a substantial project for the future, and that it is valuable to first explain and understand the theoretical prediction, as is done here.  Those analysing actual data can of course look for this pattern without needing an already implemented test.
\end{quote}

\begin{quote}
Rev. 1 is also concerned that only under dominance is simulated.  However, it is clear that any form of selection that gives the same marginal selection coefficient will give essentially the same effect: a large body of work on clines shows that different selective mechanisms are indistinguishable, and it is reasonable for the authors to rely on that.  The other main concern was about the novelty of Eqs 4, 5; these are indeed standard (which I think the authors intended by referring to them as the Kolmogorov backward equations).  I don't know of a reference where they have been written in exactly this form, but the recombination terms and the diffusion terms in Eq. 4 both go back a long way.
\end{quote}

\begin{quote}
Reviewer 2 is mainly concerned about the fit between theory and simulations.  I am not worried about this, since the deterministic mathematics stands independently, and the numerical computations are straightforward.  The stochastic simulations are useful mainly as illustrations of the process.  Theory and simulations are compared in the supplementary material; it should be stated more clearly how closely these match. (A precise fit is not expected anyway because there is substantial drift).  Rev. 2 also worries that the simulations are over a rather narrow range; I suspect that the authors should have results that address sensitivity to the boundaries.  Regarding specific point 1, I don't know of a previous derivation of exactly these equations, but similar equations have been used widely, and I don't think that the authors suggest otherwise.
\end{quote}

\noindent \hspace{4em}
\begin{minipage}{3in}
\noindent
{\bf Sincerely,}

\vskip 2em

{\bf 
Alisa Sedghifar,
Yaniv Brandvain, and
Peter Ralph
}\\
\end{minipage}

\vskip 4em


%%%%%%%%%%%%%%
\reviewersection{AE}


\begin{quote}
This is a nice analysis of how neutral genome introgresses past a cline maintained by selection against heterozygotes.  There has been a good deal of work on related problems, but mostly assuming equilibrium, and mostly on the forwards evolution of allele frequencies, rather than tracing back the ancestry of blocks of genome.  The theory presented here is therefore new, and directly relevant to interpreting genomes sampled from hybrid zones.
\end{quote}

\begin{point}{Fig 1 caption}
``alelle''
\end{point}

\reply
Fixed.

\begin{point}{\revref}
``$s$ and $\sigma$ are small'' - but relative to what?  This is not at all obvious.
\end{point}

\reply
We just meant that in the rescaling argument for a discrete model this is required formally,
without giving it enough thought.
We've added more thoughtful discussion.

\begin{point}{\revref}
Surely if time is rescaled as well as distance and genome length, everything should look the same?  
All the results, and to a good approximation the simulations, are in the continuous regime where we can rescale.
\end{point}

\reply
Good point; we in fact outlined this rescaling at \llname{ll:rescaling}.
It still seems fair to say there is ``more information'' here, because we can pick up the difference in time scaling
using haplotypes and linkage;
we've edited this paragraph to make the point better.

\begin{point}{}
A comment on two dimensions would be helpful.  
Ancestral lineages simply diffuse transversely at a rate $\sigma$, independent of their motion parallel to the cline.
\end{point}

\reply
We've added a short note on this (with more references) \revref.

\begin{point}{\revref}
This is a discrete deme model, so dispersal can't follow a continuous Gaussian. 
I guess what is happening is that $\sigma$ is chosen larger relative to deme spacing.  
However, there is some approximation here, and the approach will fail if $\sigma$ is small relatve to deme spacing.
\end{point}

\reply
There might have been some confusion about our use of the floor function?
We now describe the dispersal distribution in a more straightforward way \revref.
It's a good point more generally that if $\sigma$ is smaller than the deme spacing this scheme won't get dispersal with $\sigma$ close to the $\sigma$ that we plug into the Gaussian;
it's for this reason that we compute numerically what $\sigma$ actually is when comparing simulation to theory.

\begin{point}{}
Missing semicolon?
\end{point}

\reply
Fixed.

\begin{point}{}
``are have''
\end{point}

\reply
Fixed.

\begin{point}{\revref}
``reality'' means ``simulation'' here??
\end{point}

\reply
Well, we hoped to mean ``reality'', but perhaps ``verify that theory matches reality'' is too strong; we've adjusted this to say ``realistic, individual-based models.''.
%\alisa{I think we do mean ``reality" here, but alternate opinions welcome}
%\plr{I changed to ``realistic, individual-based models''.}

\begin{point}{}
 Some appendices might be transferred to supplementary material -- the derivations are fairly standard.
\end{point}

\reply
We're not strongly against moving the appendices to supplementary material,
but we'd prefer to keep them as appendices (and thus distributed with the document)
because although the results are fairly standard, 
they might not be familiar to our readership.  
We think it is also useful to have a more detailed description of a concrete model
that falls in the domain of attraction that we discuss here.


%%%%%%%%%%%%%%
\reviewersection{1}


\begin{quote}
    This manuscript presents a discussion of patterns of linkage and allele frequency clines in a hybrid zone between two populations, under a scenario of mild hybrid incompatibility. It presents deterministic equations for the distribution of ancestral haplotypes along the genome and argues that the lengths of ancestry segments can be used to infer selection.

    The idea of using the lengths of ancestry blocks to detect selection in hybrid zones is interesting, and I had not encountered it before. This article will likely be of most interest to theorists, however, as the discussion of the practical confounders of the selection test is limited. 
\end{quote}

\begin{point}{\revref}
    A serious limitation of the test is that it only detects selection against heterozygotes. They claim that this would not affect results, [previously at] l.453-456:
    \textit{``Although our assumed scenario of selection against heterozygotes at a single locus is uncommon in nature, we believe that this is unlikely to drastically influence our findings, as previous studies have demonstrated that underdominant loci share similar properties with more realistic models''}
    is unconvincing.
\end{point}

\reply
We had failed to cite the substantial literature on the matter (referred to by the AE);
we have now added pointers to this in the introduction \llname{ll:all_clines_are_the_same} and to the referenced section \revref.
The point is made quite well by \citet{Slatkin1973}: the details of the shape are determined mostly by the mean fitness of heterozygotes,
a parameter that collapses many models (approximately) onto each other,
at least after drift and environmental heterogeneity is considered.

\begin{point}{}
    From a theoretical perspective, the article makes modest but valid contributions, and  Equation 4 and 5 are potentially interesting extensions to equation (1). 
    The practical benefit of using these equations rather than forward simulations appears limited.
\end{point}

\reply
We also pondered this point, and feel that the equations contribute substantially
to building intuitive understanding for how to address haplotype lengths,
and aim to present analytical expressions and simulations so that they each support the conclusions of the other.
We believe that analytical results can allow for a deeper understanding than simulation output,
as for instance the intuitive discussion at \llname{ll:discussion_summary},
which mostly derives from the analytical model.
To aid our readers in developing this understanding, we have gone through considerable pains to map terms in our equations to the biological processes of migration, selection and recombination. 
We believe that our efforts will help build a better mechanistic understanding in our readers in addition to making them aware of our results.

Furthermore, as noted at \llname{ll:ld_comps}, the equations can be used to generate expectations for LD much more efficiently than forward-time simulation.

\begin{point}{}
    While reading the manuscript, I found it difficult to distinguish the most important observations from more tangential observations. The manuscript reads like a long list of observations, and would benefit from more focus on the main findings and more structure.
\end{point}

\reply
\yb{
    We agree that a long set of theoretical results can be tedious. We have therefore reshaped the introduction to highlight our major question ``How does selection against incompatibilities in a hybrid zone influence the length distribution of ancestry tracts?'' 
}

\begin{point}{}
Since the recombination-based terms in Eqs 4 and 5 do not depend on the spatially explicit part of the model or on the selection term, I would be surprised if they had not been derived somewhere.
\end{point}

\reply
We agree that this representation has likely been derived many times. 
We turned to Hartl and Clark and found a similar derivation, which is now cited in this point in the text.

\begin{point}{Eq.\ref{eqn:g_pde}}
If $a<b<\theta<0$, then I interpret $g(\theta,b) = 0$. If correct, this should be specified.  Alternately, the bounds of the integral could be modified to account for this situation.
\end{point}

\reply
Thanks for catching the mistake; we modified the bounds of the integral.

\begin{point}{}
As an example of a section that felt unstructured, consider the discussion starting at \revref.  This long paragraph starts with a long list of observations, without giving a hint as to where we are going. Beginning the paragraph with the main idea for the paragraph would be useful.
\end{point}

\reply
\plr{(comment is more general than that particular location; say something about this)}
We added an introductory sentence at \revref to provide some structure. 
We have also gone through to clarify other sections that felt unclear.

\begin{point}{\revref}
``In the notation above, $q_z(x, t, r) = P^x{Z_t = A}$.''
I did not see the notation $P^x$ above, and the RHS does not depend on small z.
\end{point}

\reply
Good point; we've used a more common notation to specify that $X_0=x$ (and $Z_0=z$ now).

\begin{point}{} 
``As we show below the solutions provide an good approximation''
\end{point}

\reply
Fixed.

\begin{point}{\revref}
It seems like the x+r and x-r should have the same sign.
\end{point}

\reply
That's correct; we've fixed the sign issues.

\begin{point}{\revref}
$g_z(x, T, a,b)$ is defined as: \textit{``We are interested in the probability $g_z(x, t; a, b)$ of finding an entire segment $(a, b)$ of ancestry $A$ in an individual having selected allele of type $z$, \ldots ''}
This definition does not make it clear that the probability is conditional on $z$; $\P(A|z)$ vs $\P(A,z)$.
\end{point}

\reply
This was indeed confusing; it's rewritten to be hopefully less confusing.

\begin{point}{l.606}
Reference issues
\end{point}

\reply
\alisa{This is a reference that is expected to appear in the same issue, so we do not have page numbers. The hope is that this can be filled in during final proofs for the issue}



% Quality of Science: Experimentally and/or theoretically excellent reliable data, no fatal flaws
% 
% Importance of Science: Research addresses a consequential question in ecology, evolution, behaviour, or conservation
% 
% Quality of Presentation: Ideas and methods mostly clear, but grammar and/or spelling is poor, format does not follow guidelines, and /or there is redundancy between sections
% 
% Does this manuscript require significant reduction in length? If 'Yes', please indicate where shortening is required in your specific comments: Yes
% 
% Does the Data Accessibility section list all the datasets needed to recreate the results in the manuscript? If `No', please specify which additional data are needed in your comments to the authors.: Yes


%%%%%%%%%%%%%%
\reviewersection{2}

\begin{quote}
    In this paper, the authors extend previous work on the diffusion of neutral alleles in a secondary-contact zone to include selection against hybrids. Building on the classic diffusion model for the dynamics of a cline at a selected locus, they formalize the dynamics of linked neutral loci and investigate the distribution of ancestry block-lengths around the selected locus.

    The authors obtain numerical results for the flattening of the cline at a neutral locus linked to an under-dominant selected locus in the absence of genetic drift. They also outline scaling arguments for the rates of movement and length of the introgressing blocks around the selected locus. Simulations are set up to test the theoretical predictions and illustrate other interesting patterns such as the distribution of haplotype (ancestry block?) lengths.
\end{quote}

\begin{point}{}
    Unfortunately, the fit of the theoretical predictions to simulations is largely not shown, and those comparisons in the manuscript are limited to the frequency dynamics of the neutral alleles (Figure S3). The predictive theoretical results are based on scaling arguments - but it is not clear which of these results are uniquely derived in this paper.
\end{point}

\reply
\alisa{Our hope was that the simulations and theory would provide qualitative support for one another. Some predictions made by the theory e.g. time and spatial scales do seem to correspond. We will try to do a better job of signposting our unique contributions??}

The rescaling arguments we provide  are of a standard sort in the literature;
we've inserted comments to this effect, at \llname{ll:rescaling} and \llname{ll:may_citation}.


\begin{point}{}
    The authors also aim to characterize the movement of the genomic block that contains a selected locus through a hybrid zone. Yet, the size of the simulated habitat barely covers the cline width, and the simulation results may be sensitive to the boundaries.

Overall, this MS presents some interesting insights, and the formalization of the block dynamics can be useful. However, I think that more work is needed to explain and test (by simulations) the novel predictions. Currently, the theory and the matching simulations are largely presented separately.
\end{point}

\reply
\alisa{Should we be running more sims? Nick seems to think we have done this, but I don't think we have.}
\plr{run sims with longer range}
\plr{also run sims in part of param range where nothing's supposed to happen?}


\begin{point}{Figure 1}
 Plotting the haplotypes to the side of the plot would help.
\end{point}

\reply
\alisa{We have remade Figure 1, with additional information that we hope will help with interpretation. We have also related this figure to the terms in Equation 5.}


\begin{point}{\revref}
 As the theory is set up in the absence of genetic drift, and, correspondingly, the local population size is large, it is unclear how a deme with 500 individuals could become empty in a Wright-Fisher model.  The (narrow) width of the habitat also needs justification.
\end{point}

\reply
Although most of the simulations shown in the main text have high population densities,
we have run simulations with lower densities, 
as shown for instance in figures \ref{sfig:condAlleleFreq_tau80_comparison}, \ref{sfig:condAlleleFreq_tau320_comparison}, and \ref{sfig:condAlleleFreq_tau1280_comparison}.
Demes are never empty for very large populations, but
we did encounter empty demes in some simulations with smaller population sizes.
We don't examine the parameter regime with neighborhood size on the order of ten individuals,
where we expect different dynamics to be at play,
and have a note to this effect \llname{ll:much_lower_densities}.

As for the width of the range, we had indeed also run additional simulations (and numerical solutions) on wider ranges to check for boundary effects;
we now have a note to this effect at \llname{ll:narrow_sims}.

\begin{point}{Figure 3:}
What are the replicate lines for the same $r$? Why re the lines for $r = 0$ all slightly displaced to the right? Also, it would appear that the color scheme in associated Figure S2 (bottom) is mismatched with the legend, shifted to the red part of the spectrum. After fixing this, I think that the Figure S2 should be in the main text instead of Fig. 3.
\end{point}

\reply
\alisa{We have switched figures S2 and 3 and fixed the colors. The colored lines are not all represented in the legend which seems to have caused some confusion and we have tried to address this in the figure legends. We don't interpret the mismatch as a displacement to the right, as translating the lines horizontally will not improve the overall fit, and attribute this to drift in the simulations.}

\begin{point}{\revref}
Total population size is irrelevant when predicting the effects of genetic drift (besides, some simulations only have 50 demes across, not 100). In a one-dimensional habitat, one uses $N \sigma \sqrt{s}$ to judge whether the effect of genetic drift is likely to be negligible.
\end{point}

\reply
\plr{We meant that the allele frequencies in the \emph{entire population} had shifted, which is determined by the total (effective) population size.  
However, we decied to }
% \yb{Thanks for spotting this. Because we do nothing quantitative with respect to drift here, we}
\yb{simply state the our observed wobble of the center of the tension zone is consistent with previous theory. However, more broadly throughout the manuscript we are not sure about relying on $N \sigma \sqrt{s}$ as we are concerned with loci linked to the selected site, and we are not aware of results extending this rule to an unlinked locus.}
\plr{(Do you mean "linked locus"?)}
%\alisa{We are concerned here with linked loci, so the $N\sigma\sqrt{s}$ rule of thumb does not necessarily apply here? Also, within demes, population size is much smaller (500?)}
%\plr{I agree: they are probably thinking of selected loci, not linked ones. We're talking about global drift not local drift. Yaniv to investigate.}

\begin{point}{Scaling arguments, p. 14, 15:}
 Please give a citation with the known scaling, and take more care with the novel ones.  
\end{point}

\reply
As noted above, we have provided some citations to similar arguments in the literature
(although such arguments are often used without citation; see, for instance, \citet{durrett2007width}).

\begin{point}{\revref}
 I think the sentence about the maximum length of the enclosing haplotype, $\sigma \sqrt{s} / x$ is mis-phrased: presumably this should be a statement on the distribution (expected size?) of the introgressing block length at distance $x$, as a lucky block can travel many sigmas in one generation?
\end{point}

\reply
Good point, it was meant to be a statement about an order of magnitude bound to a stochastic quantity rather than a precise bound; it's been rephrased.

\begin{point}{Discussion:}
    It is important to make the reader aware that even neutral clines do not just keep flattening -- genetic drift leads to a finite cline width even for neutral clines. Despite the occasional mention of genetic drift, the MS does not clearly state that the parameter space is chosen such that genetic drift is weak and does not indicate when this is a reasonable assumption (see comment above [about \revreffull{2}{6}]).
\end{point}

\reply
We now have much more discussion about drift, in particular at \revref.

\plr{I don't understand this:}
\alisa{We should add a line or two in the discussion about drift. However, IBD is perhaps not relevant over the timescales we are considering and is not necessarily ancestry informative, and since we are not interested in the equilibrium state, it is not clear that we need to worry about this limit. }

\begin{point}{rates at \revref, e.g. Eq. (7,8):}
 I find the explanation of the rates a bit unclear. I understand that the recombinant AB can arise either by A reproducing at rate $s_A$ and recombining with B, or B reproducing at rate $s_B$ and recombining with A -- such that the neutral locus of the type A is kept. Or alternatively, the BA recombinant arises, where the neutral locus at B is kept. Which one of the two happens has a probability of 1/2.
\end{point}

\reply
Right; we've added more discussion of this.

\begin{point}{}
The ``Data Accessibility section'' is missing. The folder at \href{http://github.com_petrelharp_clinal-lineages}  appears to be a working folder with lots of other files and the r-scripts hidden under a `sim' directory. The README file appears insufficient to allow orientation in the scripts, and worryingly, it has an `oh, wrong recombination distance' note to it.
\end{point}

\reply
\alisa{We have created a new repository with minimally required scripts and a descriptive README file}

\begin{point}{}
The Acknowledgements section is missing.
\end{point}

\reply This has been added. Thanks.


% Quality of Science: Mostly competent, but suffering from flaws of a technical or analytical nature
% 
% Importance of Science: Research addresses a consequential question in ecology, evolution, behaviour, or conservation
% 
% Quality of Presentation: Writing is clear, methods and data analyses are transparent, ideas make sense, proper grammar and spelling is employed, redundancy is avoided
% 
% Does this manuscript require significant reduction in length? If 'Yes", please indicate where shortening is required in your specific comments: No
% 
% Does the Data Accessibility section list all the datasets needed to recreate the results in the manuscript? If `No', please specify which additional data are needed in your comments to the authors.: No


% \end{document}
