% \documentclass[12pt,titlepage]{article}
% 
% %% THE USEPACKAGES NECESSARY FOR THIS EXAMPLE
% %% NOTE THAT genetics_manu_style MUST BE CALLED AFTER mychicago
% \usepackage{graphicx}
% %\usepackage{figcaps}
% \usepackage[nomarkers,notablist,nofiglist,tablesfirst]{endfloat}
% \usepackage{amsfonts}
% \usepackage{sectsty}
% %\usepackage{subfigure}
% %\usepackage{subcaption}
% \usepackage{natbib} \bibpunct{(}{)}{;}{author-year}{}{,} 
% \usepackage[usenames,dvipsnames,svgnames,table]{xcolor}
% %\allsectionsfont{\sffamily}
% %% THE MANUSCRIPT TITLE
% 
% \usepackage{caption}
% \usepackage[labelformat=simple]{subcaption}
% 
% \usepackage{bibentry}
% 
% \usepackage{xr}
% \externaldocument{hybrid_zone}
% 
% \newcommand{\alisa}[1]{{\em \color{red} #1}}
% \newcommand{\plr}[1]{{\em \color{blue} #1}}
% \newcommand{\yb}[1]{{\em \color{magenta} #1}}
% 
% 
% \newcommand{\given}{\,\vert\,}
% \newcommand{\st}{\,\colon\,}
% \renewcommand{\and}{\,\&\,}
% 
% 
% \begin{document}
% \setcounter{table}{0}
% \renewcommand{\thetable}{S\arabic{table}}
% \setcounter{figure}{0}
% \renewcommand{\thefigure}{S\arabic{figure}}

\begin{figure}
    \includegraphics{figs/linked-frequencies-longtime}
    \caption{
        \textbf{Probabilities of $A$ ancestry,}
        across space (vertical axis, in units of $\sigma$) 
        and time (horizontal axis, in generations).
        In each plot, color corresponds to the expected frequency of $A$ ancestry
        at a particular location in time and space.
        The selection coefficient is $s=.02$.
        \textbf{Top left:} at the selected site, showing establishment and stabilization of the cline
        on a time scale of $1/s=50$ generations.
        \textbf{Bottom left:} at an unliked site,
        with cline flattening continuing with $\sqrt{t}$.
        Remaining figures show frequencies of $A$ ancestry \emph{conditional}
        on the ancestry at the selected site,
        at different distances from the selected site ($r=.01$, .04, and 0.5 Morgans),
        as described in the text (see definition of $q_z(x,t,r)$).
        See figure \ref{fig:linked_cline_heatmaps}
        for the same figure over a shorter period of time.
    }
    \label{sfig:linked_cline_heatmaps_longtime}
\end{figure}




\begin{figure}
    \includegraphics{figs/alleleFrequencies_sim_comparison}
    \caption{
        \textbf{Comparison of simulated to theoretical clines:}
        As in Figure \ref{alleleFreq_tau100},
        frequency of ancestry B,
        across geography and at several physical positions on the genome,
        at $T=100$ generations after secondary contact,
        and at $s=0.1$.
        Dotted lines are expected frequencies of $B$ ancestry $1-q(x,t,r)$
        (unconditional on the ancestry of the linked allele), computed numerically.
        Each line represents a locus some distance $r$ away from the true target of selection, 
        and $r=0$ represents the locus that is under selection. Grey lines represent the same positions from a simulation with
        identical parameters except $s=0$.
    }\label{sfig:alleleFreq_tau100_comparison}
\end{figure}


\begin{figure}
\includegraphics[width=\textwidth]{figs/plot_chromosomes_tau100}
\caption{Randomly sampled chromosomes  across a hybrid zone of age $\tau=100$. Here we compare chromosomes of length 1M from a neutral zone to one that has a single under-dominant locus ($s=0.1$) in the middle of the chromosome (indicated by black arrow). Red blocks along the chromosome denote ancestry $B$, and orange blocks are ancestry $A$. Regions surrounding the selected site are failing to introgress.
 }\label{Fig:resistanceToIntrogression100g}
\end{figure}


\begin{figure}
\includegraphics{figs/blocksAlongChromHeatmapAncBConditioning}
\caption{Heatmap of mean block length $l_B(m_i)$ along a simulated chromosome. }\label{Supp:blockLengthHeatmap}
\end{figure}




\begin{figure}
\includegraphics{figs/blocksAlongChromNoConditioning}
\caption{Heatmap of mean block length $l_\bullet(m_i)$ surrounding a given position along the genome with a single underdominant site ($s=0.01, \tau=1000$). }\label{Supp:blockLengthNoAnc}
\end{figure}

\begin{figure}
\includegraphics{figs/blocksAlongChromHeatmap}
\caption{Heatmap of mean block length $l_\bullet(m_i)$ along a simulated chromosome with a single underdominant site ($s=0.01, \tau=1000$) }\label{Supp:blockLengthHeatmapNoAnc}
\end{figure}


\begin{figure}
\includegraphics{figs/adjacentBlocksAlongChromAncBConditioning}
\caption{Mean block length of $l_A(m_i\cdot)$ across chromosome with single under dominant site ($s=0.01, \tau=1000$)}\label{Supp:adjacentBlocks}
\end{figure}


\begin{figure}
\includegraphics{figs/ratioAdjacentBlocksAlongChromAncBConditioning}
\caption{Ratio $\frac{2\sum{l_B(m_i)}}{\sum{l_A(m_i-)+I_A(m_i+)}}$ of mean block length and adjacent block lengths across a simulated chromosome with a single underdominant site ($s=0.01, \tau=1000$).  Each line represents a deme and is normalized by mean block length across the chromosome in the deme.}\label{Supp:ratioBlockAdjacent}
\end{figure}


\begin{figure}
\includegraphics{figs/ratioAdjacentBlocksAlongChromHeatmapAncBConditioning}
\caption{Heatmap of $\frac{2\sum{l_B(m_i)}}{\sum{l_A(m_i-)+I_A(m_i+)}}$ across a simulated chromosome with a single underdominant site ($s=0.01, \tau=1000$). }\label{Supp:ratioBlockAdjacentHeatmap}
\end{figure}


\begin{figure}
    \includegraphics{figs/cond_freqs_comparison_tau_80}
    \caption{
        \textbf{Conditional frequencies of ancestry $A$ at $\tau=80$},
        comparing simulation and theory, with $\sigma=3$ deme spacings, $s=0.05$, and 50 individuals per deme.
        The top left figure shows observed and expected genotype frequencies for the two homozygotes and the heterozygote at the selected locus;
        expected genotype counts were obtained assuming random mating,
        and by solving equation \eqref{eqn:cline_pde} numerically.
        The remaining figures show observed and expected frequencies of $A$ ancestry,
        separately conditioned on the identity of the linked allele at the selected site.
        Observed frequencies become much noisier where the linked allele becomes rarer.
    } \label{sfig:condAlleleFreq_tau80_comparison}
\end{figure}


\begin{figure}
    \includegraphics{figs/cond_freqs_comparison_tau_320}
    \caption{
        \textbf{Conditional frequencies of ancestry $A$ at $\tau=320$},
        as in figure \ref{sfig:condAlleleFreq_tau80_comparison}.
        Deviations are larger than at $\tau=80$, 
        due to genetic drift.
    } \label{sfig:condAlleleFreq_tau320_comparison}
\end{figure}


\begin{figure}
    \includegraphics{figs/cond_freqs_comparison_tau_1280}
    \caption{
        \textbf{Conditional frequencies of ancestry $A$ at $\tau=1280$},
        as in figure \ref{sfig:condAlleleFreq_tau80_comparison}.
        Deviations are larger still than at $\tau=320$, 
        due to genetic drift.
    } \label{sfig:condAlleleFreq_tau1280_comparison}
\end{figure}



% \end{document}
