---
title: "Comparison of simulation to theory"
date: "`r date()`"
---

```{r doc_setup, include=FALSE}
fig.dim <- 5
knitr::opts_chunk$set(fig.height=fig.dim,fig.width=2*fig.dim,fig.align='center')
source("sim-fns.R")
library(ReacTran)
library(Matrix)
source("../notes/tran.1D.R",chdir=TRUE) # fix for log.A
source("../notes/cline-fns.R",chdir=TRUE)
```

Here's the things that we'd like to have plots of simulation compared to theory for,
and might put in the paper:

1.  **Establishment of the selected cline:** the frequency of the selected allele, $p(x,t)$,
    across space, at a few times.

2.  **Conditional ancestry probabilities:** the conditional frequency of ancestry $A$
    given being linked to a selected allele of ancestry $z$, denoted $q_z(x,t,r)$,
    across space, at a few times or distances and for each $z$.

3.  **Relaxation of linked clines:** the frequency of ancestry $A$, denoted $p(x,t,r)$, 
    across space, at a few times and one or two recombination distances. 
    (This comes after $q$ because it is computed using $q$.)

4.  **Haplotype probabilities:** the probability of seeing a haplotype of length at least $\ell$
    covering position $r$ on the genome at location $x$ in space: for fixed $\ell$,
    against $r$ and/or $x$.

*Note:* This can be run as a script, by specifying `simchunks.file` first,
and then rendering this document, e.g.:
```
Rscript -e "library(templater); simchunks.file='simulation_SIGMA1_Ninds5000_ndemes100_s0.1_dir/tau20_dir/results_simsums_chunks.Robj'; render_template('compare-to-theory.Rmd',output='simulation_SIGMA1_Ninds5000_ndemes100_s0.1_dir/tau20_dir/results_comparison-to-theory.html')"
```


## Parameters from a simulation

We'll use some simulation results produced by [yaniv_sims.R](yaniv_sims.R),
and need to convert those parameters to those from theory.
Dispersal in the simulation moves as $\lfloor Z + 1/2 \rfloor$,
where $Z \sim N(0,\sigma^2_\text{demes})$;
we'll use this fact to compute the actual $\sigma$.
(Yes, that's an unbiased walk, thanks to the floor.)
Migrants past the endpoints stay there (they don't reflect).
We'll work in spatial units equal to one deme spacing;
since generations are nonoverlapping, one generation is indeed one time step in the simulation.

```{r sim_results}
if (!exists("simchunks.file")) { 
    simchunks.file <- "simulation_SIGMA1_Ninds25000_ndemes50_s0.1_tau100_run2016-01-11_simsums_chunks.Robj" 
}
load(simchunks.file)  # provides 'chunks' and 'chunks.rvals'
get_simparams <- function (fname) {
    strings <- c( sigma="SIGMA", ninds="Ninds", ndemes="ndemes", s="s", tau="tau" )
    lapply( strings, function (x) { as.numeric(gsub(sprintf(".*[/_]%s([0-9.]+)_.*",x),"\\1",fname)) } )
}
sim.params <- get_simparams( simchunks.file )
# recombination distances
# sim.rvals <- c(0,0.5-0.01,0.5)
sim.chroms <- sapply( chunks.rvals, "[", 1 )
sim.rvals <- ifelse( sim.chroms==1, sapply( chunks.rvals, "[", 2 ) - 0.5, 0.5 )
# spatial locations of the demes
xx <- seq_len(sim.params$ndemes)-(sim.params$ndemes+1)/2
theory.params <- list( sigma=trueSigma(sim.params$sigma), s=sim.params$s, tau=sim.params$tau, 
       density=sim.params$ninds/sim.params$ndemes,
       width=sim.params$ndemes )
simchunks.file
unlist(theory.params)
```

*Note:* We should also work out what the actual value of **s** is.

Here are some functions we'll use to extract what we want
from data in the `chunks` format:
```{r helper_fns}
source("chunks_fns.R")
```


# Selected cline

To get the value predicted from theory, we'll solve the corresponding PDE:
```{r selected_cline_theory}
tt <- seq(0,theory.params$tau,length.out=21)
xgrid <- setup.grid.1D(x.up=-theory.params$width/2, x.down=theory.params$width/2, N=2*theory.params$width)
fgrid <- extend_grid(xgrid)
fwds.soln <- forwards_pde(s=theory.params$s, times=tt, grid=fgrid, sigma=theory.params$sigma )
```
... and, compare it to the observed values:
```{r selected_cline_sim}
gcounts <- genotype_counts(chunks[[1]])
# sim frequencies
sfreqs <- sweep( tcrossprod(gcounts,dip.to.hap), 1, 2*rowSums(gcounts), "/" )
# theory frequencies
tfreqs <- cbind( A=fwds.soln[nrow(fwds.soln),-1], B=1-fwds.soln[nrow(fwds.soln),-1] )
matplot(xx, sfreqs, type='l', lty=1, main=sprintf("Selected allele frequency, t=%d",sim.params$tau), ylab="frequency", xlab='geographic position')
matlines(fgrid$x.mid, tfreqs, lty=2, lwd=2)
legend("topright", legend=outer(colnames(sfreqs),c("sim","theory"),paste,sep=" : "), lty=rep(1:2,each=2), col=rep(1:ncol(sfreqs),2))
```

For fun, let's also look at genotype frequencies (is there HW?):
```{r selected_geno_freqs}
matplot(xx, sweep(gcounts,1,rowSums(gcounts),"/"), lty=1, type='l', main=sprintf("Selected genotype frequencies, t=%d",sim.params$tau), ylab="frequency", xlab='geographic position')
matlines(fgrid$x.mid, cbind(AA=tfreqs[,"A"]^2,AB=2*tfreqs[,"A"]*tfreqs[,"B"],BB=tfreqs[,"B"]^2), lty=4, lwd=2)
```
Looks pretty similar, just shifted.

**Note:** Verified by looking at simulations with small $\tau$,
the shift in cline center is due to the cline drifting.


# Ancestry probabilities

First, compute theoretical predictions at the values of $r$ in the simulation:
```{r linked_soln}
linked.solns <- lapply( sim.rvals, function (r) {
        forwards_backwards_pde(s=theory.params$s, times=tt, grid=xgrid, r=r, 
                           sigma=theory.params$sigma, fwds.soln=fwds.soln, log.p=TRUE)
    } )
```
... and compare to the observed values:
```{r compare_linked}
for (kr in seq_along(sim.rvals)[-1]) {
    gcounts <- conditional_genotype_counts(chunks[[kr]], chunks[[1]])
    # frequencies from simulation
    cfreqs <- conditional_freqs(gcounts)
    # frequencies from theory
    tfreqs <- linked.solns[[kr]][nrow(linked.solns[[kr]]),-1]
    dim(tfreqs) <- c( length(xgrid$x.mid), 2 )
    matplot(xx, cfreqs, type='l', lty=1, 
            main=sprintf("Frequency of A ancestry by selected background, t=%d, r=%0.3f",sim.params$tau, sim.rvals[kr]), 
            ylab="frequency", xlab='geographic position')
    matlines(xgrid$x.mid, tfreqs, lty=2, lwd=2)
    legend("topright", legend=outer(colnames(cfreqs),c("sim","theory"),paste,sep=" : "), lty=rep(1:2,each=2), col=rep(1:ncol(sfreqs),2))
}
```

To help interpret this, here's the raw counts of the four phased haplotypes across the range:
(labels are "linked-selected", for instance, "B-A" has B ancestry linked to the selected A allele)
```{r hap_counts}
for (kr in seq_along(sim.rvals)[-1]) {
    hcounts <- conditional_haplotype_counts( conditional_genotype_counts(chunks[[kr]], chunks[[1]]) )
    hc.names <- outer( dimnames(hcounts)[[2]], dimnames(hcounts)[[3]], paste, sep="-")
    dim(hcounts) <- c( dim(hcounts)[1], prod(dim(hcounts)[2:3]) )
    colnames(hcounts) <- hc.names
    matplot(xx, hcounts, type='l', 
            main=sprintf("Haplotype counts (linked-selected), t=%d, r=%0.3f", sim.params$tau, sim.rvals[kr]), 
            ylab="count", xlab="geographic position")
    legend('topleft', lty=1:5, col=1:6, legend=hc.names )
}
```


# Linked clines

This will just be a combination of the previous results,
since it is local proportions of each type without conditioning on the allele at the selected locus.
```{r linked_clines}
for (kr in seq_along(sim.rvals)[-1]) {
    gcounts <- genotype_counts(chunks[[kr]])
    # sim frequencies
    sfreqs <- sweep( tcrossprod(gcounts,dip.to.hap), 1, 2*rowSums(gcounts), "/" )
    # theory frequencies
    cline.soln <- cline_from_soln(linked.solns[[kr]],grid=xgrid)
    tfreqs <- cbind( A=cline.soln[nrow(cline.soln),-1], B=1-cline.soln[nrow(cline.soln),-1] )
    matplot(xx, sfreqs, type='l', lty=1, main=sprintf("Linked allele frequency, t=%d, r=%0.3f",sim.params$tau, sim.rvals[kr]), 
            ylab="frequency", xlab='geographic position')
    matlines(xgrid$x.mid, tfreqs, lty=2, lwd=2)
    legend("topright", legend=outer(colnames(sfreqs),c("sim","theory"),paste,sep=" : "), lty=rep(1:2,each=2), col=rep(1:ncol(sfreqs),2))
}
```


# Haplotype probabilities

The probability of seeing a haplotype of length at least $\ell$
covering position $r$ on the genome at location $x$ in space: for fixed $\ell$,
against $r$ and/or $x$.

We'll do this against $x$ at a few values of $r$ because that's what we have.
(Don't mess with this next code chunk, it'll invalidate the cache, which was expensive to make.)
```{r theory_haps, cache=TRUE}
rgrid <- setup.grid.1D(x.up=-0.5, x.down=0.5, N=41)
hap.soln <- forwards_backwards_haplotypes(s=theory.params$s, times=tt, xgrid=xgrid, rgrid=rgrid,
                           sigma=theory.params$sigma, 
                           fwds.grid=fgrid, fwds.soln=fwds.soln )
```

```{r get_mean_hap_lens}
hap.pdf <- hap_cdf_to_pdf( hap.soln )
```

Now let's compare this to simulation:
```{r sim_hap_lens}
sim_lens <- sapply( chunks, sapply, mean )
sim_cond_lens <- sapply( chunks, sapply, function (x) { if (any(x>0)) { mean(x[x>0]) } else { NA } } )

hap.len.mat <- sapply( chunks.rvals, function (crv) { 
               haplens <- mean_hap_len( hap.pdf, loc=if(crv[1]==1){crv[2]-0.5}else{max(rgrid$x.mid)} ) 
               haplens[length(tt),1+seq_along(xgrid$x.mid)]
           } )

matplot(xx, sim_lens, type='l', lty=sapply(chunks.rvals,"[",1),
        main="mean haplotype length", ylim=range(sim_lens,hap.len.mat),
        xlab="spatial position", ylab="mean B haplotype length (incl. zeros)")
matlines( rev(xgrid$x.mid), hap.len.mat, lty=3 )
legend("topleft", lty=c(sapply(chunks.rvals,"[",1),3), col=c(rep_len(1:6,ncol(sim_cond_lens)),1),
        legend=c( sapply( chunks.rvals, function (x) { sprintf("chr %d: %0.3f",x[1],x[2]) } ), "theory") )
```

Here are haplotype length distributions:
```{r hap_len_distributions}
hap.coords <- expand.grid( x=xgrid$x.mid, AB=c("A","B"), rk=1:nrow(attr(hap.pdf,"r")) )
hap.coords$p <- hap.pdf[nrow(hap.pdf),-1]
hap.coords$cump <- hap.soln[nrow(hap.pdf),-1]
hap.coords$right <- attr(hap.pdf,"r")[hap.coords$rk,"right"]
hap.coords$left <- attr(hap.pdf,"r")[hap.coords$rk,"left"]
hap.coords$len <- hap.coords$right-hap.coords$left

x.inds <- floor(seq(1,length(xgrid$x.mid),length.out=11))
xx <- xgrid$x.mid[x.inds]
nxab <- attr(hap.pdf,"soln.dims")[1]  # number of space x AB combinations
r.lens <- attr(hap.pdf,"r")[,"right"] - attr(hap.pdf,"r")[,"left"]
lenvals <- seq(0,1,length.out=21)
hap.distrns <- sapply( x.inds, function (xk) {
        kk <- 1 + which( ( (0:(ncol(hap.pdf)-2)) %% nxab )+1 == xk )
        y <- tapply( hap.pdf[nrow(hap.pdf),kk], match(r.lens,sort(unique(r.lens))), sum )
        # not really a density: need to divide
        diff(c(0,approx( sort(unique(r.lens)), cumsum(y), xout=lenvals )$y))
    } )
matplot( lenvals, hap.distrns, type='l', xlab="haplotype length", ylab="probability density" )
legend("topright", lty=1:5, col=1:6, legend=sprintf("location = %0.2f", xx))
```

Here are haplotype probabilities (**not** cumulative):
```{r look_at_haps}
for (haplen in c(0.02, 0.04, 0.08, 0.16, 0.32, 0.64) ) {
    hap.probs <- get_hap_probs( hap.pdf, left=-haplen/2, right=haplen/2 )
    image(hap.probs, legend=TRUE, yaxt='n', zlim=range(hap.probs[,-1]),
       main=paste(c("A:","B:"),paste(round(attr(hap.probs,"r"),3),collapse=" to ")) )
    axis(2,at=(pretty(xgrid$x.mid)-min(xgrid$x.mid))/diff(range(xgrid$x.mid)),labels=pretty(xgrid$x.mid))
}
```
and here are the same offset by `r rgrid$dx[1]`
```{r look_at_haps_offset}
for (haplen in c(0.02, 0.04, 0.08, 0.16, 0.32, 0.64) ) {
    hap.probs <- get_hap_probs( hap.pdf, left=rgrid$dx[1], right=rgrid$dx[1]+haplen )
    image(hap.probs, legend=TRUE, yaxt='n', zlim=range(hap.probs[,-1]),
       main=paste(c("A:","B:"),paste(round(attr(hap.probs,"r"),3),collapse=" to ")) )
    axis(2,at=(pretty(xgrid$x.mid)-min(xgrid$x.mid))/diff(range(xgrid$x.mid)),labels=pretty(xgrid$x.mid))
}
```

Here are also heatmaps of **mean lengths**:
```{r hap_lens}
hap.centers <- c(0.0, 0.02, 0.04, 0.08, 0.16, 0.32)
for (kloc in seq_along(hap.centers)) {
    hap.len <- mean_hap_len( hap.pdf, loc=hap.centers[kloc] )
    image(hap.len, legend=TRUE, yaxt='n', zlim=range(hap.len[,-1]),
          main=paste(c("A:","B:"),sprintf("mean haplotype length at %f",hap.centers[kloc])) )
    axis(2,at=(pretty(xgrid$x.mid)-min(xgrid$x.mid))/diff(range(xgrid$x.mid)),labels=pretty(xgrid$x.mid))
}
```
