---
title: "Comparison of simulation to theory"
date: "`r date()`"
---

```{r doc_setup, include=FALSE}
fig.dim <- 5
knitr::opts_chunk$set(fig.height=fig.dim,fig.width=2*fig.dim,fig.align='center')
source("sim-fns.R")
```

Here's the things that we'd like to have plots of simulation compared to theory for,
and might put in the paper:

1.  **Establishment of the selected cline:** the frequency of the selected allele, $p(x,t)$,
    across space, at a few times.

2.  **Conditional ancestry probabilities:** the conditional frequency of ancestry $A$
    given being linked to a selected allele of ancestry $z$, denoted $q_z(x,t,r)$,
    across space, at a few times or distances and for each $z$.

3.  **Relaxation of linked clines:** the frequency of ancestry $A$, denoted $p(x,t,r)$, 
    across space, at a few times and one or two recombination distances. 
    (This comes after $q$ because it is computed using $q$.)

4.  **Haplotype probabilities:** the probability of seeing a haplotype of length at least $\ell$
    covering position $r$ on the genome at location $x$ in space: for fixed $\ell$,
    against $r$ and/or $x$.



# Parameters from a simulation

We'll use some simulation results produced by [yaniv_sims.R](yaniv_sims.R),
and need to convert those parameters to those from theory.
Dispersal in the simulation moves as $\lfloor Z + 1/2 \rfloor$,
where $Z \sim N(0,\sigma^2_\text{demes})$;
we'll use this fact to compute the actual $\sigma$.
(Yes, that's an unbiased walk, thanks to the floor.)
Migrants past the endpoints stay there (they don't reflect).

```{r sim_results}
simchunks.file <- "simulation_SIGMA1_Ninds25000_ndemes50_s0.1_tau1000_simsums_chunks.Robj"
sim.params <- list( sigma=1, ninds=25000, ndemes=50, s=0.1, tau=1000 )

theory.params <- list( sigma=trueSigma(sim.params$sigma), s=sim.params$s, tau=sim.params$tau )
```

