# Important files in this directory

Here's the files that you might want to pay attention to.

- [sims-fns.R](sims-fns.R) contains the utility functions for running simulations and plotting results.
	
- [run_sims.R](run_sims.R) will run the simulation; see below for how to use it.
	
	It will create files in the format 	`simulation_SIGMA%s_Ninds%s_ndemes%s_s%s_dir/tau%s_dir/results_runid_%s`,
	including some useful plots. 	
	



# To run a simulation 

Edit `run_sims.R` and execute it.
The parameters are at the top of the file:
```
params = list(
    SIGMA = 3,        # sigma for dispersal, in deme spacings
    ndemes = 100,     # number of demes
    deme_size = 50,   # number of diploid individuals per deme
    S = 0.05,         # selection coefficient, s
    zone_age = 10     # number of generations to run simulation for
)
```

It will create files in the format 	`simulation_SIGMA%s_Ninds%s_ndemes%s_s%s_dir/tau%s_dir/results_runid_%s`,
(runid is a unique identifier automatically generated by the script)
including some useful plots.
To extract parameters back out of the file names, see the `get_simparams` function below.


## Description of resulting .Robj files

The simulation saves out results as R objects, as follows (with run-specific string as above prepended):

* `_params.Robj` : records the parameters used for this run

* `_rawoutput.Robj` : contains a single object, `sims.full`, that is a list with one entry per value of `zone_age`;
        each element has the following elements (`[[zone_age]][[c("inds","sp.inds","deme","pars","QTL")]]`):

    - `"inds"` is a list of length `ninds`, one for each individual (`[[ind]][[chr]][[X1/X2]]`):

        * each element is a list of length 2 (the number of chromsomes), named "chr1" and "chr2":

            * each element of these is a list of length 2 (the ploidy), named "X1" and "X2"

                * each element of these is a data frame, with columns

                    - `pos` :  coordinates of the beginning of the genomic segment
                    - `ancest` :  ID of the ancestor in the original generation from whom the following segment was inherited from

    - `"sp.inds"` is a character vector of length `ninds`, giving the ancestral types of original ancestors (`k`th entry is type of ancestor with ID `k`)
    - `"deme"` is an integer vector of length `ninds`, giving the original deme number of each ancestor
    - `"pars"` is an integer, named `n.gen`, that gives the number of generatsion of the simulation
    - `"QTL"` is the QTL vector used in the simulation (see `run_sims.R` for its self-explanatory format)

* `_simsums.Robj` : contains a single object, `sims.sums`, which is a list: `[[zone_age]][[c("ind.ancest","ancest.prop")]]`

    * `"ind.ancest"` is a list of length `ninds` (`[[ind][[chr]]][[X1/X2]]`), with one entry per individual, each element being:

        * each element is a list of length 2 (the number of chromsomes), named "chr1" and "chr2":

            * each element of these is a list of length 2 (the ploidy), named "X1" and "X2"

                * each element of these is a data frame, with columns 
                    
                    - `starts` : the start of a contiguous chunk of ancestry
                    - `stops` : the end of that chunk (will be the same as the next `starts`)
                    - `sp2` : TRUE or FALSE depending if that chunk inherits from "species 2" of the original ancestors
        
    * `"ancest.prop"` is a `ninds` by `ploidy` matrix, with one row per individual and one column for each copy of chromosome 1 
        that gives the ancestry proportion of each individual on each of their two copies of chromosome 1.

* `_simsums_chunks.Robj` : has a single object, "chunks", which is a list:

    * with components:

        - `selected` : at the selected site
        - `distant` : at recombination distance 0.01
        - `unlinked` : at the center of the other chromosome

    * and each of these is a list with one component per deme, each of which is

        - a matrix of chunk sizes, of dimensions `deme_size` by `ploidy`, with columns named e.g. "X1" and "X2" for the chromosomes.
        - Each matrix entry gives the chunk size of Ancestry B surrounding the locus (0 means the locus is of Ancestry A)

    The script could be modified to record these summaries at other locations.


# Plots of the results

The plots that are produced by `run_sims.R` are (with the run-specific string prepended as above):

- `freqplot.pdf` : plot of allele frequencies across space at specified loci
- `chunks_ecdf.pdf` : estimated cumulative distribution function of chunk sizes for each deme
- `chunks_density.pdf` : distribution of chunk sizes
- `chunks_mean.pdf` : chunk lengths across simulated chromosome
- `LD.pdf`


The plots produced by `haplotype-plots.R`, run on the output (see below) are:

- `ratioAdjacentBlocksAlongChromAncBConditioning_nonnormalized.pdf`
- `blocksAlongChromNoConditioning_nonnormalized.pdf`
- `blocksAlongChromAncBConditioning_nonnormalized.pdf`
- `adjacentBlocksAlongChromNoConditioning_nonnormalized.pdf`
- `adjacentBlocksAlongChromAncBConditioning_nonnormalized.pdf`
- `ratioAdjacentBlocksAlongChromNoConditioning.pdf`
- `ratioAdjacentBlocksAlongChromHeatmapNoConditioning.pdf`
- `ratioAdjacentBlocksAlongChromHeatmapAncBConditioning.pdf`
- `ratioAdjacentBlocksAlongChromAncBConditioning.pdf`
- `blocksAlongChromNoConditioning.pdf`
- `blocksAlongChromHeatmap.pdf`
- `blocksAlongChromHeatmapAncBConditioning.pdf`
- `blocksAlongChromAncBConditioning.pdf`
- `adjacentBlocksAlongChromNoConditioning.pdf`
- `adjacentBlocksAlongChromAncBConditioning.pdf`

## Making these plots

These plots can be produced by `haplotype-plots.R`;
this runs on the `sims.sums` object saved out by the script in the file `results_runid_%s_simsums.Robj`.
The script `make-plots.R` will load the required things and run `haplotype-plots.R`;
it works as follows:
```
get_simparams <- function (fname) {
    strings <- c( SIGMA="SIGMA", ninds="Ninds", ndemes="ndemes", S="s", tau="tau", run.id="runid_" )
    out <- lapply( strings, function (x) { as.numeric(gsub(sprintf(".*[/_]%s([0-9.]+)_.*",x),"\\1",fname)) } )
    out$zone_age <- out$tau
    out$deme_size <- out$ninds/out$ndemes
    return(out)
}

simdir <- "simulation_SIGMA3_Ninds250000_ndemes500_s0.05_dir/tau100_dir"
simsum.files <- list.files(simdir, ".*simsums.Robj", full.names=TRUE)
sapply( simsum.files, function (fname) {
        params <- get_simparams(fname)
        load(fname)
        source("haplotype-plots.R")
    } )
```

This script also saves out some time-consuming intermediate steps, namely:

	- `intervalSizes.Robj` :
	- `intervalSizes_allAncs.Robj` :
	- `flanking.blocks.by.ind.Robj` : 
	- `flanking.blocks.by.ind.AncB.Robj` :


## To compile the resulting figures across time points at the same parameter values

Since each set of parameter values gets a directory,
with time points in subdirectories,
we compile all the figures of each sort into multipage pdfs to put in the parameter value directory.
This is done with `compile-pdfs.sh`: 
change to the directory corresponding to parameter values, then run `compile-pdfs.sh`, e.g.:
```
cd simulation_SIGMA3_Ninds6000_ndemes120_s0.001_dir
../compile-pdfs.sh
```
This creates compilation files, e.g. `density.pdf`, within the simulation directory.

## Making plots with `potential_statistics.R` :

This script will generate (**how?**) the following plots and R object files:

	* `statistic_length.pdf` : mean length across ALL indivduals in the zone (including non-recombinant chroms) vs chromosome position
	* `statistic_ecdf.pdf` : ECDF of length (i.e. prob that random haplotype is longer than this one) relative to ALL individuals in a given population.
	* `statistic_length_of_adjacent_blocks.pdf` : mean length of blocks adjacent to the block containing the selected locus.

# Finding numbers of distinct ancestors by location

To get the family size of lineages, run `number_of_inds.R`, editing the file directly to specify the simulation to be analyzed. 
This will generate an R object file **with what?**


# To create a "comparison to theory" document for a given simulation

There is an Rmarkdown file, `compare-to-theory.Rmd`, that can be used to create plots
comparing the results of a given simulation
to predictions from numerical integration of the PDEs,
using the tools in [notes/](../notes/).

To create it,
install the minimal package [templater](https://github.com/petrelharp/templater) (e.g. by `devtools::install_github("petrelharp/templater")`)
define `simchunks.file`, and then compile the Rmarkdown file, for instance:
```
library(templater)
simchunks.file <- "simulation_SIGMA1_Ninds5000_ndemes100_s0.1_dir/tau500_dir/results_simsums_chunks.Robj"
output.file <- gsub("_simsums_chunks.Robj", "_comparison-to-theory.html", simchunks.file)
render_template('compare-to-theory.Rmd',output=output.file)
```

If you have `make`, you can do:
```
make simulation_SIGMA1_Ninds5000_ndemes100_s0.1_dir/tau500_dir/results_comparison-to-theory.html
```
or, to run it for all simulation results in a directory, do
```
SIMDIR=simulation_SIGMA1_Ninds5000_ndemes100_s0.1_dir/
for x in $(find $SIMDIR -name "*_simsums_chunks.Robj")
do 
    y=$(echo $x | sed -e 's/_simsums_chunks.Robj/_comparison-to-theory.html/') 
    make $y
done
```

