---
title: "Running simulations by Alisa"
author: Alisa
date: "`r date()`"
output: html_document
---

```{r doc_setup, include=FALSE}
fig.dim <- 5
knitr::opts_chunk$set(fig.height=fig.dim,fig.width=2*fig.dim,fig.align='center')
source("sim-fns.R")
```

Setting up the simulation
-------------------------
To simulate, we read in files from sim-fns.R
```{r load_fns}
	source("sim-fns.R")
```

Start off with simulation of 50 generation zone with 50 demes of 500 individuals each and sigma of 1:
```{r set_deme_parameters}
	zone_age = 50
	ndemes = 50
	deme_size = 500
	ninds = ndemes*deme_size
	SIGMA = 1
```
note that zone_age above can be a vector of times.

Now to define selection. Here we selection against heterozygotes to be strong (0.5).
```{r set_selection}
	zone_age = c(50)
	qtl=list(chr1=data.frame(traits=c("underdominant"), s = S,pos=c(0.5)))
```
`qtl` is a named list, one element per chromosome,
and each element of the list is a data frame
whose columns give the name of the locus,
the selection coefficient,
and the position in Morgans.

Running the simulation
-------------------------

This is a two-step process. First we simulate genotypes:
```{r run_sim_step1}
	sims.s0.1 = lapply(zone_age,sim.zone,n.ind=ninds,n.deme=ndemes,sigma=SIGMA)
```
The output is: a list with hierarchy: [[zone_age]][[c("inds","sp.inds","deme","pars","QTL")]]
	where:
		"inds" is a list of length ninds, such that [[ind]][[chr]][[X1/X2]][[data.frame where col1 = coordinates of rec., col2 = ancestor ID]]
		"sp.inds" is a character vector of length ninds, giving the ancestry given an ancestor ID
		"deme" is an integer vector of length ninds, giving the deme number of an ancestor, given their ancestor ID
		"pars" is an integer that gives the number of generatsion of the simulation
		"QTL" returns the QTL vector used in the simulation

The second step is to parse the simulated genotypes:
```{r run_sim_step2}
	sims.sums = lapply(sims.s0.1,spBreaks)
```
The output here is a list: [[zone_age]][[c("ind.ancestry","ancest.prop")]]
		"ind.ancestry" is a list of length ninds: [[ind][[chr]]][[X1/X2]][[data.frame where col1 = (chunk)starts, col2 = (chunk)stops, col3 = sp2(T/F)]]
		"ancest.prop" is a ninds*2 matrix giving the ancestry proportion of each individual on each of their two copies of the chromosome.
		
This resulting files can get quite large, but we've saved them here
```{r save_file}
	save(sims.sums,file="~/Projects/HybridZones/ClineProjects/sims_sums_s0.5_sigma1_50*500_tau50.Robj")
```


Getting relevant statistics
-------------------------
Let's start by defining a vector of loci at which we will measure things:

```{r focal_loci}
	loci = list(seq(0.5,1,0.05))
```

Now, we get frequencies at each of the sites in the above vector, 
first by getting genotypes at all loci, then using those to calculate an allele frequency for each deme:
```{r get_frequencies}
	genotypes = lapply(sims.sums,function(Z){data.frame(do.call(cbind,lapply(Z$ind.ancest, geno.ind, loci)))})
	freqs = lapply(my.genos,function(X){apply(X,1,function(Z){tapply(Z,cut(1:ncol(X),breaks=seq(0,ncol(X),deme_size)),mean)/2})})

		matplot(xx,freqs[[1]],col=rainbow(20),lty=1,type="l", main="50 generations",ylab="freq",xlab="deme")
		legend('bottomright',col=rainbow(20),lty=1,legend=loci[[1]],cex=0.75)
		matpoints(-xx,pp,col=rainbow(20),lty=3,type="l")

	
```
(`genotypes` is a list of nloci*ninds dataframes where each entry is the number of ancestry B alleles in an individual at the given locus.)
Each line in the plot represents a locus a given distance from the selected locus.

We can compare these frequencies to predictions made by theory :

```{r peters_predictions}
	zeta <- function(x,r){
	2*pnorm(abs(x)) - 1 + 2 * exp(r*abs(x) + r^2/2 + pnorm(abs(x)+r,lower.tail=FALSE,log.p=TRUE))	
	}

	pcline <- function(x,r,tau=zone_age,sigma=SIGMA){
	(1/2)*(1-sign(x)*zeta(x=x/sqrt(tau*sigma^2),r=r*tau))	
	}

	rr=(loci[[1]]-0.5)*SIGMA/sqrt(S) # loci at which frequency is measured
	xx <- (1:ndemes)-0.5-ndemes/2	 # deme positions
	pp <- do.call(cbind,lapply(rr,function(R){pcline(xx,r=R,tau=zone_age,sigma=SIGMA)})) #allele frequency


		matplot(xx,freqs[[1]],col=rainbow(20),lty=1,type="l", main="50 generations",ylab="freq",xlab="deme")
		legend('bottomright',col=rainbow(20),lty=1,legend=loci[[1]],cex=0.75)
		matpoints(-xx,pp,col=rainbow(20),lty=3,type="l")

```

Next, we think about tract lengths in these populations


First we need to decide where the loci under selection are.
Here `qtl` is a named list, one element per chromosome,
and each element of the list is a data frame
whose columns give the name of the locus,
the selection coefficient,
and the position in Morgans.

```{r setup_qtl}
qtl=list(
         chr1=data.frame(
                    traits = c("underdominant"),
                    s = c(0.1),
                    pos = c(0.5)
                )
    )
```

Now, let's set up the population,
which means specifying the number of individuals,
the number of demes,
the width of the zone in real units,
and the dispersal SD ($\sigma$).
Our locus has a selection coefficient of `r qtl[[1]]$s`,
so the cline width will be about `r round(1/sqrt(qtl[[1]]$s),digits=1)`$\sigma$ wide;
so we'll take our zone to be four times that,
and measure distance in units of $\sigma$:
```{r setup_pop}
zone.width <- ceiling(4/sqrt(qtl[[1]]$s))  # in real units
n.ind <- 1000
n.deme <- 50
sigma <- 1 # in real units
deme.spacing <- zone.width/n.deme
deme.locs <- (seq_len(n.deme)-0.5-(n.deme/2))*deme.spacing
sigma.demes <- sigma/deme.spacing # in units of demes
```
The discretized dispersal kernel moves as $\lfloor Z+1/2 \rfloor$,
where $Z \sim N(0,\sigma_\text{demes}^2)$;
let's check this has SD close to $\sigma$:
```{r check_sigma}
zz <- floor( rnorm( 1e6, mean=1/2, sd=sigma.demes ) )*deme.spacing
sd(zz)
```
That's pretty close.



Running a simulation
--------------------

The simulation doesn't know how we're measuring things in real units,
so the sigma we pass needs to be in units of demes.
It returns all the information about the individuals,
but then we use `spBreaks()` to extract which individual is which ancestry.
```{r run_quick_sim}
sim <- sim.zone( n.ind=n.ind, n.deme=n.deme, n.gen=10, sigma=sigma.demes, quiet=TRUE )
sim.sums <- spBreaks(sim)
```

OK, now let's genotype these individuals at some loci:
```{r genotype_sim, fig.width=3*fig.dim, fig.height=1.5*fig.dim}
loci <- seq(0,1,0.05)
loci.cols <- rainbow(12)[cut(abs(loci-0.5),breaks=min((length(loci)-1)/2,8))]
info <- data.frame(
               chr=1,
               bp=floor(loci*1e6),
               pos=loci
           )
genotypes <- geno.pop( sim.sums$ind.ancest, list(chr1=loci) )
```

Now, let's
compute ancestry frequencies,
and plot the resulting clines:
```{r genotype_freqs, fig.width=3*fig.dim, fig.height=1.5*fig.dim}
freqs <- do.call( cbind, tapply( 1:ncol(genotypes), sim$deme, function (kk) {
                    rowMeans(genotypes[,kk])
                } ) )
matplot(deme.locs, t(freqs), type='l', col=loci.cols, xlab="spatial position", ylab="ancestry frequency")
legend("topleft",lty=1,legend=loci,col=loci.cols,title="locus position")
```

We can also look at the local frequencies of the diploid genotypes at each locus:
```{r diploid_freqs}
diploid.freqs <- lapply( seq_along(loci), function (locus) {
                do.call( rbind, tapply( 1:ncol(genotypes), sim$deme, function (kk) {
                    as.vector( tabulate( 1+genotypes[locus,kk], nbins=3 ) )
                } ) )
           } )
# one end
matplot(diploid.freqs[[1]],type='l',main=paste("locus at",loci[1]),lty=1,col=c('red','black','green'))
legend('topright',lty=1,col=c('red','black','green'),legend=c("homozygote","heterozygote","homozygote"))
# the selected locus
sel.locus <- match( qtl$chr1$pos, loci )
matplot(diploid.freqs[[sel.locus]],type='l',main=paste("selected locus at",loci[sel.locus]),lty=1,col=c('red','black','green'))
legend('topright',lty=1,col=c('red','black','green'),legend=c("homozygote","heterozygote","homozygote"))
# the other end
matplot(diploid.freqs[[length(loci)]],type='l',main=paste("locus at",loci[length(loci)]),lty=1,col=c('red','black','green'))
legend('topright',lty=1,col=c('red','black','green'),legend=c("homozygote","heterozygote","homozygote"))
```
