---
title: "Diffusion local times and the time until recombination"
date: "`r date()`"
---

$$
\newcommand{\E}{\mathbb{E}}
\renewcommand{\P}{\mathbb{P}}
\newcommand{\calE}{\mathcal{E}}
\DeclareMathOperator{\sgn}{sgn}
\newcommand{\grad}{\nabla}
$$

```{r doc_setup, include=FALSE}
fig.dim <- 5
knitr::opts_chunk$set(fig.width=2*fig.dim,fig.height=fig.dim,fig.align='center')
library(ReacTran)
```

*Useful functions used below:*
```{r setup_fns}
source("cline-fns.R")
```

The goal
========

The key quantity we compute with diffusion theory 
is the probability that a lineage, 
begun at location $x$ and killed at rate $k$ when in heterozygotes,
is still alive at time $\tau$.
Denote this by
$$
\begin{aligned}
    \zeta(x,k,\tau).
\end{aligned}
$$
Now, let $q(x,r,\tau,A)$ denote the probability that a lineage begun at $x<0$
at a location at recombination distance $r$ from the selected site
inherits from the left-hand side of the barrier
(i.e., an ancestor on $(-\infty,0)$)
$\tau$ generations ago.
The probability that a recombination occurs 
between the lineage and the selected locus on either pair of chromatids
is $2r$,
and if this happens, the lineage ends up on either background with probability $1/2$.
At this point, its probability of originally coming from one or the other background
still depends on its location,
but if we **assume this is 1/2**,
for an individual to the left of the barrier ($x<0$),
the probability of inheriting from the left of the barrier at $\tau$
is the probability that no recombination at the barrier has occurred,
plus one-half the probability that recombination in a heterozygote has occurred,
so if $\zeta = \zeta(x,2r,\tau)$, then
$q=\zeta +(1-\zeta)/2=(1+\zeta)/2$.
If $x>0$ this is the one-half times probability a recombination in a heterozygote has occurred,
i.e., $q=(1-\zeta)/2$.
Combining these,
the probability of ending up on the left-hand side is *approximated* as
$$
\begin{aligned}
    q(x,r,\tau,A) = 
    \frac{1}{2} \left\{ 1 - \sgn(x) \zeta(x,2r,\tau) \right\} .
\end{aligned}
$$



Approximations, generally
=========================

We have a function $p(x)$ (the local proportion of $A$ alleles),
and another function $f_H(x)$ (the local probability an $A$ is in a heterozygote),
which happens to equal $(1-s)(1-p(x))/(1-s(1-p(x)))$ but we might not want to use that fact;
then $p$ defines a diffusion $X$ by $dX_t = \log(p(x))' dt + dB_t$
and then we define $H_t = \int_0^t f_H(X_s) ds$.
We want to know $\E^x[\exp(-r H_t)]$ as a function of $x$, $r$, and $t$.

More concretely,
we want the distribution of the time to first recombination in a heterozygote;
denote this by $T$, so that
$$
\begin{aligned}
\P\{ T > t \} 
&= \exp\left\{ - \int_0^t r f_H(X_s) ds \right\} \\
&= \exp\left\{ - \int_{-\infty}^{\infty} r f_H(x) \ell_t(x) dx \right\}  ,
\end{aligned}
$$
where $\ell$ is the local time density for $X$.
This is the Laplace transform of 
the heterozygote-weighted local time, $H_t = \int f_H(x) \ell_t(x) dx$,
so it means that we want to come up with a reasonable approximation to this quantity.

This suggests two possible approximations:

1. A lineage doesn't go past the hybrid zone,
   and only recombines when it's in the hybrid zone,
   which has width something,
   and where the probability of being in a heterozygote is 1/2.

2. A lineage spends little time in the hybrid zone,
   and recombines when it's past the hybrid zone,
   where it's type is rare,
   thus replacing both the drift $\partial_x \log(p(x))$
   and the probability of being in a heterozygote $f_H(x)$ with step functions.

The first one seems more natural,
but it requires two parameters: 
what the width should be and where the barrier should be.


The right model
===============

When the lineage is at $x$, 
the lineage recombines to the other background
in a given generation
with probability equal to the recombination rate 
multiplied by the probability of being found in a heterozygote.
If $p_{AB}(x)$ is the local fraction of heterozygotes,
and heterozygotes have relative fecundity $1-s$,
then the probability of being in a heterozygote is
$(1-s)p_{AB}(x)/((1-s)p_{AB}(x)+2p_{AA}(x))$;
if diploid genotypes are assorted randomly,
then $p_{AB} = 2p(1-p)$ and $p_{AA}=p^2$, so the probability that an $A$ lineage
passes through a heterozygote in a given generation is
$$
\begin{align*}
f_H(x) 
&= 
\frac{ (1-s)p(x)(1-p(x)) }{ p(x)^2 + (1-s) p(x) (1-p(x)) }
&= 
\frac{ (1-s)(1-p(x)) }{ p(x) + (1-s) (1-p(x)) } \\
&= 
\frac{ (1-s)(1-p(x)) }{ 1 - s (1-p(x)) } .
\end{align*}
$$

Bazykin's approximation to a cline is
$$
p(x) = \frac{1}{2} (1+\tanh(-2x\sqrt{s}/\sigma))
$$
Here's what that looks like:
```{r tanh_cline}
tanh_cline
s <- 0.1
xx <- seq(-15,15,length.out=500)
plot(xx, (-1)*tanh_drift(xx,s=s), type='l', col='red', xlab="space", ylab='proportion or drift')
lines(xx, tanh_cline(xx,s=s), col='black')
lines(xx, 2*tanh_cline(xx,s=s)*(1-tanh_cline(xx,s=s)), col='blue')
lines(xx, tanh_het(xx,s=s), col='green')
legend("topleft",lty=1,col=c("black","red","blue","green"),
        legend=c("p(x)","-log(p(x))'","2p(1-p)","prob of het"))
```

A lineage does something a bit more complicated,
but let's just run a random walk with drift equal to the logarithmic derivative of $p(x)$,
i.e.,
$$
\begin{align*}
    \partial_x \log(p(x)) 
    &= -\frac{2\sqrt{s}}{\sigma} 
        \frac{1-\tanh^2(-2x\sqrt{s}/σ)}{1+\tanh(-2x\sqrt{s}/σ)} .
\end{align*}
$$
(Really, the drift should be the logarithmic derivative of $p(x)(1-sf_H(x))$.)

```{r sim_lineage,cache=TRUE}
tanh_drift
nwalks <- 1e4
ntimes <- 4e3
s <- 0.1
tau <- 5
xx <- seq(-15,15,length.out=500)
tt <- seq(0,tau,length.out=ntimes)
rws <- sim_diffusion( tanh_drift, tau, nwalks, ntimes, s=s, xinit=rnorm(nwalks) )
layout( t(1:2), widths=c(3,1) )
matplot(tt, rws[,1:80],type='l', ylab='position', xlab='time', ylim=range(rws) )
legend("topright",lwd=2,legend=c("prob of het","cline","drift"),col=c("black","green","red"))
plot( (-1)*tanh_drift(xx,s=s), xx, col='red', lwd=2, type='l', ylim=range(rws), ylab='', xlab='drift or prob' )
lines( tanh_cline(xx,s=s), xx, col='green', lwd=2 )
lines( (1-s)*(1-tanh_cline(xx,s=s))/(1-s*(1-tanh_cline(xx,s=s))), xx, lwd=2 )
```
The cline is in green and the local probability of being in a heterozygote is in black;
the drift (multiplied by $-1$) is in red.

Now let's get local times for those walks.
```{r lineage_local_times,cache=TRUE,dependson="sim_lineage"}
xx <- seq( min(rws), max(rws), length.out=200 )
xmids <- xx[-1]-diff(xx)/2
lts <- attr(rws,"dt")*apply( matrix(findInterval(rws,xx),nrow=nrow(rws)), 2, tabulate, nbins=length(xx)-1 )
layout( t(1:2), widths=c(3,1) )
matplot(lts[,1:30],xmids,type='l',xlab="local time", ylab="space")
plot( tanh_cline(xx,s=s), xx, col='green', lwd=2, type='l', ylim=range(rws), ylab='' )
lines( (1-s)*(1-tanh_cline(xx,s=s))/(1-s*(1-tanh_cline(xx,s=s))), xx, lwd=2 )
```


Here is the distribution of $T=\int_0^{`r tau`} f_H(X_t) dt$:
```{r dist_het_time, cache=TRUE, dependson="sim_het_time"}
het.time <- weighted_time( rws, tanh_het, s=s )
layout(1:2)
hist( het.time, breaks=50, xlab="time in hets", freq=FALSE, main="histogram of T" ) 
hist( log(het.time), breaks=50, xlab="log(time in hets)", freq=FALSE , main="histogram of log(T)") 
```

The exact solution
------------------

Let
$$
\zeta(x,t,r) = \P\{ T_r > t \} .
$$
Then the Kolmogorov forward equation tells us that $\zeta$ solves the following differential equation:
$$\begin{aligned}
    \frac{\partial}{\partial t} \zeta(x,t,r) 
    &= 
    \frac{\sigma^2}{2} \frac{\partial^2}{\partial x^2} \zeta(x,t,r) + \frac{\partial }{\partial x} \log(p(x)) \frac{\partial }{\partial x} \zeta(x,t,r)  - r f_H(x) \zeta(x,t,r)  \\
    \zeta(x,0,r) = 1 .
\end{aligned}$$

So, let's solve that, numerically.
We'll use the `ReacTran` library,
which in 1D solves equations with advection-diffusion parts of the form:
$$\begin{aligned}
   \frac{\partial C}{\partial t} 
   &= 
   \frac{1}{A} \grad \cdot A \left( D \grad C - v C \right) \\
   &= 
   \left( \frac{1}{A} \grad AD - v \right) \cdot \grad C 
   - \frac{1}{A} \grad A \cdot v C 
   + D \Delta C \\
\end{aligned}$$
which is the same as what we have if we let $D=\sigma^2/2$ and $A=p$ and $v=0$.
Here is the setup:
```{r solve_pde_setup, cache=TRUE}
thegrid <- setup.grid.1D(x.up=-15, x.down=15, N = 5000)
clinefn <- setup.prop.1D(grid=thegrid, func=tanh_cline, s=s)
hetfn <- setup.prop.1D(grid=thegrid, func=tanh_het, s=s)
yinit <- rep(1.0,thegrid$N)
tt <- seq(0,tau,length.out=100)
r <- 1
```

First, two quick checks: First, by setting $f_H=1$,
we should have the solution $u(x,t,r) = \exp(-tr)$.
Then, by setting $\sigma$ to be very small, the solution at $x$ should be $\exp(-trf_H(x))$.
```{r solve_pde_check, cache=TRUE, dependson="solve_pde_setup"}
layout(t(1:2))
test.hetfn <- setup.prop.1D(grid=thegrid, func=function(x){rep(1,length(x))})
pde.fn <- function (t,y,parms,...) {
    tran <- tran.1D(C=y, A=clinefn, D=1/2, flux.up=0, flux.down=0, dx=thegrid)$dC
    list( tran - r * test.hetfn$mid * y )
}
pde.soln <- ode.1D( y=yinit, times=tt, func=pde.fn, nspec=1 )
matplot(pde.soln,type='l',ylab="probability of no recomb", col='black',log='y')
lines(tt,exp(-r*tt),lwd=2,lty=2,col='red')
## no diffusion
pde.fn <- function (t,y,parms,...) {
    tran <- tran.1D(C=y, A=clinefn, D=1e-6, flux.up=0, flux.down=0, dx=thegrid)$dC
    list( tran - r * hetfn$mid * y )
}
pde.soln <- ode.1D( y=yinit, times=tt, func=pde.fn, nspec=1 )
plot.x <- sort(match(sample(thegrid$x.mid[thegrid$x.mid>-5 & thegrid$x.mid<5],12),thegrid$x.mid))
matplot(pde.soln[,plot.x],type='l',ylab="probability of no recomb", col=rainbow(length(plot.x)), lty=1 )
matpoints( exp(outer( tt, -r*hetfn$mid[plot.x], "*" )), lty=4, col=rainbow(length(plot.x)), pch=20 )
```
**Looks good.**

Now, to solve the real problem:
```{r solve_pde, cache=TRUE, dependson="solve_pde_setup"}
pde.fn <- function (t,y,parms,...) {
    tran <- tran.1D(C=y, A=clinefn, D=1/2, flux.up=0, flux.down=0, dx=thegrid)$dC
    list( tran - r * hetfn$mid * y )
}
pde.soln <- ode.1D( y=yinit, times=tt, func=pde.fn, nspec=1 ) # note FIRST COLUMN IS TIME
image(pde.soln,legend=TRUE,main="prob of no recomb by t", yaxt='n', ylab='space')
axis(2,at=(pretty(thegrid$x.mid)-min(thegrid$x.mid))/diff(range(thegrid$x.mid)),labels=pretty(thegrid$x.mid))
```

It's reassuring the solution is between 0 and 1.
Let's spot check that with the simulation.
Here it is at $t=`r tau`$:
```{r check_pde_soln, cache=TRUE}
check.t <- function (alpha) {
    t.tau <- which.min(abs(tt-alpha*tau))
    probs <- exp( - weighted_time( rws, tanh_het, s=s, use.t=1:floor(nrow(rws)*alpha) ) )
    plot(rws[1,],probs,xlab='initial location', ylab=paste('probability of no recomb by t=',alpha*tau),ylim=c(0,1))
    lines(lowess(rws[1,],probs,f=0.05),col='red',lwd=2)
    lines(thegrid$x.int,pde.soln[t.tau,],col='green',lwd=2)
    legend("topright",lwd=2,col=c("red","green"),legend=c("mean","numerical solution"))
}
check.t(1)
```
and, at other values of $t$:
```{r check_pde_soln_2, cache=TRUE, dependson="check_pde_soln"}
layout(t(1:2))
check.t(0.1)
check.t(0.2)
check.t(0.4)
check.t(0.8)
```
**Looks pretty good.**


The even more right solution
============================

Actually, the cline gets set up
in a time-inhomogeneous way.
Let's do that, by numerical solution,
then run the lineage back through it.
The forwards-time evolution of the cline,
with isotropic density and migration,
solves the PDE (Barton 1979):
$$\begin{aligned}
  \partial_t p(x,t) = \frac{\sigma^2}{2} \partial_x^2 p(x,t)+ s p(x,t)(1-p(x,t))(2p(x,t)-1) .
\end{aligned}$$
```{r fwds_pde, cache=TRUE}
thegrid <- setup.grid.1D(x.up=-15, x.down=15, N = 5000)
yinit <- ifelse( thegrid$x.mid>0, 0, 1 )
tt <- seq(0,tau,length.out=100)
fwds.pde <- function (t,y,parms,...) {
    tran <- tran.1D(C=y, D=1/2, dx=thegrid)$dC
    list( tran + s * y * (1-y) * (2*y-1) )
}
fwds.soln <- ode.1D( y=yinit, times=tt, func=fwds.pde, nspec=1 ) # note FIRST COLUMN IS TIME
image(fwds.soln,legend=TRUE,main="mean cline development", yaxt='n', ylab='space')
axis(2,at=(pretty(thegrid$x.mid)-min(thegrid$x.mid))/diff(range(thegrid$x.mid)),labels=pretty(thegrid$x.mid))
```
Now, let's run the killed lineage backwards in it:
```{r fwds_back_pde, cache=TRUE, dependson="fwds_pde"}
yinit <- rep(1.0,thegrid$N)
rev.pde <- function (t,y,parms,...) {
    p <- cline_interp(tau-t,soln=fwds.soln)
    # p <- setup.prop.1D(grid=thegrid, xy=cbind(thegrid$x.mid,cline_interp(tau-t)), interp="linear")
    tran <- tran.1D(C=y, A=p, D=1/2, flux.up=0, flux.down=0, dx=thegrid)$dC
    list( tran - r * (1-p) * y )
}
rev.soln <- ode.1D( y=yinit, times=tt, func=rev.pde, nspec=1, tcrit=tau ) # note FIRST COLUMN IS TIME
image(rev.soln,legend=TRUE,main="prob of no recomb by t", yaxt='n', ylab='space')
axis(2,at=(pretty(thegrid$x.mid)-min(thegrid$x.mid))/diff(range(thegrid$x.mid)),labels=pretty(thegrid$x.mid))
```

Let's compare the case where the cline is set up instantly or not:
```{r compare_rightness, cache=TRUE,dependson=c("fwds_back_pde","solve_pde_setup")}
layout(t(1:2))
fb.p <- clineplot(rev.soln,main="forwards-backwards")
st.p <- clineplot(pde.soln,main="stationary cline")
layout(1)
plot(fb.p,st.p,asp=1,pch=20,cex=0.5,xlab="forwards-backwards",ylab="stationary",col=rainbow(ncol(fb.p))[col(fb.p)])
abline(0,1,lwd=3)
```

Just forwards
-------------

An alternate route to this would have been to
solve the coupled PDE for the frequencies of the selected locus 
and the linked locus.



Comparison, by simulation
=========================

Here are distributions of $T=\int_0^{`r tau`} f_H(X_t) dt$
for two different starting points: $x=0$ and $x=-3$,
and five different models:

- **tanh:** the "correct" model;
- **reflected:** a Brownian motion reflected at zero and counting one-half time in a neighborhood of size $\sigma/\sqrt{s}$;
- **reflected tanh:** a Brownian motion reflected at zero and counting time weighted by the $\tanh$ cline;
- **pushed:** a Brownian motion pushed back at rate $4\sigma\sqrt{s}$ above zero and counting time above zero;
- **pushed tanh:** a Brownian motion pushed back at rate $4\sigma\sqrt{s}$ above zero and counting time weighted by the $\tanh$ cline.

```{r sim_general, fig.height=3*fig.dim, cache=TRUE}
nwalks <- 1e4
ntimes <- 1e3
s <- 0.1
tau <- 10
initvals <- c(0,-3)
lt.comparison <- lapply( initvals, function (xinit) { list(
        tanh=weighted_time( 
               sim_diffusion( tanh_drift,tau,nwalks,ntimes,s=s, xinit=xinit ),
               tanh_het, s=s
           ),
        reflected=weighted_time(
                -abs(sim_diffusion( function(...){0}, tau, nwalks, ntimes, xinit=xinit )),
                function (x,s) { ifelse(abs(x)<1/(2*sqrt(s)), 1/2, 0) }, s=s
            ),
        reflected_tanh=weighted_time(
                -abs(sim_diffusion( function(...){0}, tau, nwalks, ntimes, xinit=xinit )),
                tanh_het, s=s
            ),
        pushed=weighted_time(
                sim_diffusion( function(x,s){ifelse(x>0,-4*sqrt(s),0)}, tau, nwalks, ntimes, s=s, xinit=xinit ),
                function (x) { x>0 }
            ),
        pushed_tanh=weighted_time(
                sim_diffusion( function(x,s){ifelse(x>0,-4*sqrt(s),0)}, tau, nwalks, ntimes, s=s, xinit=xinit ),
                tanh_het, s=s
            )
        ) } )
layout(matrix(1:(length(lt.comparison)*length(lt.comparison[[1]])),ncol=2))
for (k in seq_along(lt.comparison)) {
    for (j in seq_along(lt.comparison[[k]])) {
        hist(lt.comparison[[k]][[j]],xlim=range(lt.comparison),main=names(lt.comparison[[k]])[j],breaks=seq(0,tau,length.out=50),xlab="time in hets")
    }
}
```

Of course, we aren't actually trying to estimate the distribution of $T$ itself,
but rather the probabilities of recombination (the Laplace transform).
Here are these probabilities, as a function of $r$
(the second plots are on a log-log scale):
```{r recomb_prob_comparisons, cache=TRUE, dependson="sim_general"}
rvals <- exp(seq(log(.001),log(0.5),length.out=400))
precombs <- lapply( lt.comparison, function (lts) {
        sapply( lts, function (lt) sapply( rvals, function (r) { mean((-1)*expm1(-r*lt)) } ) )
    } )
layout(t(1:2))
for (k in seq_along(precombs)) {
    matplot( rvals, precombs[[k]], type='l', col=1:ncol(precombs[[k]]), main=paste("x =",initvals[k]), ylim=range(precombs), ylab="prob of recomb", xlab='r', lty=1 )
    legend("topright",lty=1,col=1:ncol(precombs[[k]]),legend=colnames(precombs[[k]]))
}
for (k in seq_along(precombs)) {
    matplot( rvals, precombs[[k]], type='l', col=1:ncol(precombs[[k]]), log='xy', main=paste("x =",initvals[k]), ylim=range(precombs), ylab="prob of recomb", xlab='r', lty=1  )
    legend("bottomright",lty=1,col=1:ncol(precombs[[k]]),legend=colnames(precombs[[k]]))
}
```

Here's the same thing for a much longer time:
```{r sim_general_longer, fig.height=fig.dim, cache=TRUE}
nwalks <- 1e3
ntimes <- 1e4
s <- 0.1
tau <- 100
initvals <- c(0,-3)
lt.comparison <- lapply( initvals, function (xinit) { list(
        tanh=weighted_time( 
               sim_diffusion( tanh_drift,tau,nwalks,ntimes,s=s, xinit=xinit ),
               tanh_het, s=s
           ),
        reflected=weighted_time(
                -abs(sim_diffusion( function(...){0}, tau, nwalks, ntimes, xinit=xinit )),
                function (x,s) { ifelse(abs(x)<1/(2*sqrt(s)), 1/2, 0) }, s=s
            ),
        reflected_tanh=weighted_time(
                -abs(sim_diffusion( function(...){0}, tau, nwalks, ntimes, xinit=xinit )),
                tanh_het, s=s
            ),
        pushed=weighted_time(
                sim_diffusion( function(x,s){ifelse(x>0,-4*sqrt(s),0)}, tau, nwalks, ntimes, s=s, xinit=xinit ),
                function (x) { x>0 }
            ),
        pushed_tanh=weighted_time(
                sim_diffusion( function(x,s){ifelse(x>0,-4*sqrt(s),0)}, tau, nwalks, ntimes, s=s, xinit=xinit ),
                tanh_het, s=s
            )
        ) } )
tbreaks <- seq(0,tau,length.out=50)
Tdists <- lapply( lt.comparison, function (lts) {
        sapply( lts, function (lt) tabulate(findInterval(lt,tbreaks),nbins=length(tbreaks)-1) )
    } )
rvals <- exp(seq(log(.001),log(0.5),length.out=400))
precombs <- lapply( lt.comparison, function (lts) {
        sapply( lts, function (lt) sapply( rvals, function (r) { mean((-1)*expm1(-r*lt)) } ) )
    } )
layout(t(1:2))
for (k in seq_along(precombs)) {
    matplot( tbreaks[-1]-diff(tbreaks)/2, Tdists[[k]], type='l', col=1:ncol(Tdists[[k]]), main=paste("x =",initvals[k]), ylim=range(Tdists), xlab="time in hets", lty=1 )
    legend("bottomright",lty=1,col=1:ncol(precombs[[k]]),legend=colnames(precombs[[k]]))
}
for (k in seq_along(precombs)) {
    matplot( rvals, precombs[[k]], type='l', col=1:ncol(precombs[[k]]), main=paste("x =",initvals[k]), ylim=range(precombs), ylab="prob of recomb", xlab='r', lty=1 )
    legend("bottomright",lty=1,col=1:ncol(precombs[[k]]),legend=colnames(precombs[[k]]))
}
```




Reflected Brownian motion, killed according to local time at zero
=================================================================

Now, let's look at the analytics for the first approximation.

It states in [Takacs](https://projecteuclid.org/euclid.aoap/1177004703) that 
if $\ell(x)$ is the local time at $x$ for Brownian motion run for time 1 and begun at 0,
i.e.,
$$
\begin{aligned}
    \ell(x) = \lim_{\epsilon \to 0} \frac{1}{\epsilon}
        \text{measure}\{t \,:\, x \le B_t \le x+\epsilon, \, 0 \le t \le 1\}
\end{aligned}
$$
then
$$
\begin{aligned}
    \P\{ \ell(x) \le u \} = 2 \Phi(|x|+u)-1,
\end{aligned}
$$
where $\Phi(x)=\P\{B_1 \le x\}$ is the Gaussian CDF.
Differentiating this, $\P\{\ell \in du\} = \partial_u \P\{\ell \le u\}$, so
$$
\begin{aligned}
    \P\{ \ell(x) = 0 \} &= 2 \Phi(|x|)-1 \\
    \P\{ \ell(x) \in du \} &= 2 \phi(|x|+u) ,
\end{aligned}
$$
where $\phi$ is the Gaussian density.

Let's check this:
```{r sim_rw, cache=TRUE}
nwalks <- 1e4
ntimes <- 1e3
# want total time to be tau, so:
eps <- 1/sqrt(ntimes)
rws <- rbind(rep(0,nwalks), matrix( rnorm(nwalks*(ntimes-1))*eps, nrow=ntimes-1, ncol=nwalks ) )
rws <- apply(rws,2,cumsum)
matplot(seq(0,1,length.out=ntimes),rws[,1:20],type='l', ylab='position')
```
Here we've simulated `r nwalks` random walks for a total of `r ntimes` steps,
and scaled steps to have this be a total Brownian time of 1.
Let's check;
this should just be a standard Gaussian:
```{r final_dist}
xh <- hist(rws[ntimes,],breaks=50,main="final position")
lines( xh$mids, sum(xh$counts)*diff(pnorm(xh$breaks)) )
```

Now, let's check on the local time formula.
The number of steps the random walk spends at position $k$ on a grid with spacing $\epsilon$
and a time step of \gamma
should be $\epsilon \times \ell(k\epsilon) / \gamma$.
First, the local time at 0:
```{r local_times_0, cache=TRUE, dependson="sim_rw"}
xh <- hist(colSums(abs(rws)<=1.5*eps)/ntimes/(3*eps),main="local time at 0", breaks=30)
lines( xh$mids, sum(xh$counts)*diff( 2*pnorm(xh$breaks)-1 ), col='red' )
```
What about the local time at other locations?
Note there is an atom at 0 now
(that is depicted by the dots in the figures below),
and that the discretization could trip us up;
the colors (red/green/blue) use the (mid/upper/lower) points of the interval that we count in.
```{r local_times, cache=TRUE, dependson="sim_rw"}
layout(matrix(1:4,nrow=2))
for (loc in c(0.1, 0.2, 0.4, 0.6)) {
    lts <- colSums(abs(rws-loc)<=1.5*eps)/ntimes/(3*eps)
    zeros <- (lts==0)
    pred.zero <- 2*pnorm(abs(loc)+c(0,1.5*eps,-1.5*eps))-1
    xh <- hist(lts[!zeros],main=paste("local time at",loc), breaks=30, ylim=range(c(0,pred.zero*nwalks)))
    lines( xh$mids, sum(!zeros)*diff( 2*pnorm(abs(loc)+xh$breaks)-1 ), col='red' )
    points(rep(0,4), c(sum(zeros),pred.zero*nwalks), col=1:4, pch=20, cex=2)
    # lower bound
    lines( xh$mids, sum(!zeros)*diff( 2*pnorm(abs(loc+1.5*eps)+xh$breaks)-1 ), col=3 )
    # upper bound
    lines( xh$mids, sum(!zeros)*diff( 2*pnorm(abs(loc-1.5*eps)+xh$breaks)-1 ), col=4 )
}
```
Clearly, the discretization is still tripping us up a bit,
but it's pretty good.

Its Laplace transform
---------------------

Now suppose that something happens at rate $r$ relative to the local time at $x$
(e.g., recombination to the other background).
Then the chance that this has not happened by time 1 is
$$
\begin{aligned}
\zeta(x,r) &:=
\E\left[ e^{-r\ell(x)} \right] \\
&=
2\Phi(|x|)-1 + 
\int_0^\infty e^{-rt} 2 \phi(|x|+t) dt \\
&=
2\Phi(|x|)-1 + 
e^{r|x|} \int_{|x|}^\infty e^{-rt} 2 \phi(t) dt \\
&=
2\Phi(|x|)-1 + 
2 e^{r|x|} \int_{|x|}^\infty \frac{1}{\sqrt{2\pi}} \exp\left( - rt - t^2/2 \right) dt \\
&=
2\Phi(|x|)-1 + 
2 e^{r|x|+r^2/2} \int_{|x|}^\infty \frac{1}{\sqrt{2\pi}} \exp\left( - (t+r)^2/2 \right) dt \\
&=
2\Phi(|x|)-1 + 
2 e^{r|x|+r^2/2} \left( 1-\Phi(|x|+r) \right) .
\end{aligned}
$$
At $r=0$ this equals 1, as it should, since it's the chance that nothing happens.
For large $x$, since $1-\Phi(x) \sim e^{-x^2/2}/\sqrt{2\pi}x$,
this is approximately 
$1 + 2 \exp( r|x| + r^2/2 - (|x|+r)^2/2 )$,
which goes to 1, as it should.


Let's check that, too: why not?
```{r killing, cache=TRUE, dependson="sim_rw"}
zeta.reflected
loc <- 0.1
compare.rates <- as.data.frame( t( sapply( c(0.25,0.5,1,1.5,2,2.5,3), function (rate) {
        visits <- colSums(abs(rws-loc)<=1.5*eps)
        killed <- ( rbinom(ncol(rws),size=visits,prob=rate/ntimes/(3*eps)) == 0 )
        c( 
          rate=rate,
          simulation=mean(killed), 
          theory=zeta.reflected(loc,rate),
          SE=sqrt(mean(killed)*(1-mean(killed))/nwalks)
          )
    } ) ) )
with(compare.rates, {
         plot( rate, theory, type='l', xlab="recomb rate", ylab="prob of recomb", col='red', ylim=c(0,1) ) 
         points( rate, simulation, pch=20 )
         arrows( x0=rate, y0=simulation-2*SE, y1=simulation+2*SE, angle=90, code=3 )
    } )
```

Other values of time
--------------------

The above restricted to standard Brownian motion on one unit of time.
Really, we will want to use this across $\tau$ generations,
modeling a random walk with variance $\sigma^2$ per generation.

The key quantity we want is the probability that a Brownian motion with variance $\sigma^2$ killed at rate $r$
with respect to its local time at $x$ was still alive at time $\tau$.
Denote this by
$$
\begin{aligned}
    \zeta(x,r,\tau,\sigma).
\end{aligned}
$$
By Brownian scaling,
since $B_{ct}$ has the same distribution as $\sqrt{c}B_t$,
$$
\begin{aligned}
    \zeta(x,r,\tau,\sigma) 
    &=
    \zeta(x,r\tau,1,\sqrt{\tau}\sigma) \\
    &=
    \zeta(x/\sqrt{\tau\sigma^2},r\tau) ,
\end{aligned}
$$
where
$$
\begin{aligned}
    \zeta(x,r)
    &=
    1 - 2(1-\Phi(|x|)) + 
    2 e^{r|x|+r^2/2} \left( 1-\Phi(|x|+r) \right) .
\end{aligned}
$$



Solving the PDE
---------------

Let's check this agrees with the solution for the PDE.
According to Knight 1981 (Thm 7.4.3), $\zeta'(0) = - 2 r \zeta(0)$;
ReacTran lets you specify the 'flux' at the boundary is proportional to $\zeta(0)$,
where the flux is in our case $\sigma^2/2 \zeta'$.
First note that the solution depends on where we put the boundaries:
```{r reflected_pde_boundaries,cache=TRUE}
# NOTE: must install ReacTran from R-forge, e.g. 
#   install.packages("ReacTran",repos="http://r-forge.r-project.org")
# or else
source("tran.1D.R")
# to fix bug in using convective boundary conditions.
x.eps <- 5*mean(thegrid$dx)
solve.reflected <- function (x.up=-20) {
    thegrid <- setup.grid.1D(x.up=x.up, x.down=0, N = 5000)
    hetfn <- setup.prop.1D(grid=thegrid, func=function(x){abs(x)<x.eps})
    yinit <- rep(1.0,thegrid$N)
    tau <- 40
    tt <- seq(0,tau,length.out=20)
    r <- 1
    pde.fn <- function (t,y,parms,...) {
        # flux = zeta'(0)/2 = a.bl.down*( zeta(0) - C.down )
        tran <- tran.1D(C=y, A=1, D=1/2, dx=thegrid, C.down=0, a.bl.down=4*r)$dC
        list( tran - r * hetfn$mid * y )
    }
    soln <- ode.1D( y=yinit, times=tt, func=pde.fn, nspec=1 ) # note FIRST COLUMN IS TIME
    attr(soln,"x") <- thegrid$x.mid
    return(soln)
}
pde.soln.0 <- solve.reflected(x.up=-5)
pde.soln.1 <- solve.reflected(x.up=-8)
pde.soln.2 <- solve.reflected(x.up=-20)
clineplot(pde.soln.0,x=attr(pde.soln.0,"x"),xlim=c(-5,0))
clineplot(pde.soln.1,x=attr(pde.soln.1,"x"),add=TRUE)
clineplot(pde.soln.2,x=attr(pde.soln.2,"x"),add=TRUE)
```

And, the PDE solution doesn't agree with the simulations very well: 
clearly, we need a different boundary condition at zero.
This is broken in ReacTran; I've modified the function to fix this.
```{r reflected_pde,cache=TRUE}
thegrid <- setup.grid.1D(x.up=-20, x.down=0, N = 5000)
hetfn <- setup.prop.1D(grid=thegrid, func=function(x){abs(x)<x.eps})
yinit <- rep(1.0,thegrid$N)
tau <- 40
tt <- seq(0,tau,length.out=20)
r <- 1
pde.fn <- function (t,y,parms,...) {
    # flux = zeta'(0)/2 = a.bl.down*( zeta(0) - C.down )
    tran <- tran.1D(C=y, A=1, D=1/2, dx=thegrid, C.down=0, a.bl.down=2*r )$dC
    list( tran - r * hetfn$mid * y )
}
pde.soln <- ode.1D( y=yinit, times=tt, func=pde.fn, nspec=1 ) # note FIRST COLUMN IS TIME
ps <- clineplot(pde.soln,times=seq_along(tt),xlim=c(-8,0),legend=FALSE)
matlines( thegrid$x.mid, outer( thegrid$x.mid, tt, function(x,t) { zeta.reflected(x/sqrt(t),r*t*x.eps) }), lty=3, col=rainbow(length(tt)) )
legend("bottomleft",lty=1:2,legend=c("PDE","theory"))
```
Still not right... hm.



The associated cline shape
--------------------------

The Brownian rescaling above tells us how a cline gets wider as time progresses:
for a fixed $r$, the width grows as $\sqrt{\tau}$.
It also tells us that clines with the same $r\tau$ have the same shape,
after rescaling the width by $\sqrt{\tau}$.
Also, this tells us we only have a cline for $r$ of order at least $1/\tau$:
for smaller $r$, the locus is still completely linked to the selected locus,
and hence a step function.

More generally, here's what $\zeta$ looks like:
```{r plot_zeta, fig.height=2*fig.dim}
layout(1:2)
xvals <- seq(0,5,length.out=150)
rvals <- seq(0,10,length.out=100)
zetavals <- outer( xvals, rvals, zeta.reflected )
matplot(xvals,zetavals,type='l',xlab="x",ylab=expression(zeta),col=rainbow(length(rvals)))
lrvals <- seq(1,length(rvals),length.out=6)
legend("bottomright",lty=1,col=rainbow(length(rvals))[lrvals],legend=paste("r=",round(rvals[lrvals],digits=2)))
matplot(rvals,t(zetavals),type='l',xlab="r",ylab=expression(zeta),col=rainbow(length(xvals)))
lxvals <- seq(1,length(xvals),length.out=6)
legend("bottomright",lty=1,col=rainbow(length(xvals))[lxvals],legend=paste("x=",round(xvals[lxvals],digits=2)),bg='white')
```

Recall from above that the predicted cline is
$$
\begin{aligned}
    q(x,r,\tau,A) = 
    \frac{1}{2} \left\{ 1 - \sgn(x) \zeta(x,2r,\tau) \right\} .
\end{aligned}
$$

Here's the cline shape:
```{r show_cline, cache=TRUE}
zeta.pcline <- function (x,r) { pcline(x,r,recombfn=function(x,r,...){zeta.reflected(x=x/sqrt(tau*sigma^2),r=r*tau)},tau=1,sigma=1) }
xx <- seq(-15,15,length.out=500)
for (tau in c(5,50)) {
    plot(xx, zeta.pcline(xx,r=1), type='n', 
            xlab="distance from cline center", ylab="prob inherited from the left", 
            main=as.expression(substitute(tau==mytau,list(mytau=tau))) )
    rr <- seq(0,0.2,length.out=11)
    pp <- sapply( rr, function (r) { zeta.pcline(xx,r=r,tau=tau) } )
    matplot(xx,pp,type='l',lty=1,col=rainbow(length(rr)),add=TRUE)
    legend("topright",lty=1,col=rainbow(length(rr)),legend=paste("r=",rr))
}
```

Asymptotics
-----------

This is very steep at $x=0$.
Notice that as $x$ goes to 0,
$$
\begin{aligned}
    \lim_{x \to 0} \zeta(x,r) 
    &= 2 e^{r^2/2} (1-\Phi(r)) \\
    &= \frac{2}{r\sqrt{2\pi}} + O(r^{-3}) .
\end{aligned}
$$
so that the difference in the left and right limits is
$$
\begin{aligned}
    \lim_{x \to 0-} q(x,r) 
    -\lim_{x \to 0+} q(x,r) 
    &= 
    \frac{1}{2}+\frac{2}{r\sqrt{2\pi}} 
    - \frac{1}{2}-\frac{2}{r\sqrt{2\pi}} 
      +O(r^{-3}) \\
    &=
    \frac{4}{r\sqrt{2\pi}}
      +O(r^{-3}) 
\end{aligned}
$$

Let's also look at the slope of $q$ at $x=0$.
Since $\partial_x |x| = \sgn(x)$,
$$
\begin{aligned}
    \partial_x \zeta(x,r)
    &=
    2 \sgn(x) \left(
    \phi(|x|)
    + r e^{r|x|+r^2/2} (1-\Phi(|x|+r))
    - e^{r|x|+r^2/2} \phi(|x|+r)
      \right)
\end{aligned}
$$
and so
$$
\begin{aligned}
    -\partial_x q(x)
    &=
    \phi(|x|)
    + r e^{r|x|+r^2/2} (1-\Phi(|x|+r))
    - e^{r|x|+r^2/2} \phi(|x|+r) \\
    &\to
    \frac{1}{\sqrt{2\pi}}(1-e^{r^2/2}e^{-r^2/2})
    + r e^{r^2/2} (1-\Phi(r))  \qquad \text{as } x \to 0 \\
    &=
    r e^{r^2/2} (1-\Phi(r)) \\
    &=
    \frac{1}{\sqrt{2\pi}} + O(\frac{1}{r}) .
\end{aligned}
$$

Also note that by the asymptotics for $1-\Phi(x)$,
$$
\begin{aligned}
    \zeta(x,r) \approx 1 - \sqrt{\frac{2}{\pi}} e^{-x^2/2} \left( \frac{r}{x+r} \right) \qquad \text{for large }|x| .
\end{aligned}
$$



Brownian motion with drift above zero
=====================================

A better model for the process might be Brownian motion
that, when above zero, has negative drift,
and is killed at some rate.
In other words, 
$$\begin{aligned}
dX_t = -\theta(X_t) dt + dB_t,
\end{aligned}$$
where $\theta(x)=\theta$ for $x \ge 0$, and $\theta(x)=0$ for $x<0$.


Solving the PDE
---------------

Above we simulated the clines deriving from this process.
```{r sim_pushed, cache=TRUE}
rvals <- exp(seq(log(.001),log(0.5),length.out=10))
pushfn <- function(x,s){ifelse(x>0,-4*sqrt(s),0)}
nwalks <- 1e4
ntimes <- 1e3
s <- 0.1
tau <- 10
xinit <- seq(-8,8,length.out=nwalks)
sim.rw <- sim_diffusion( pushfn, tau, nwalks, ntimes, s=s, xinit=xinit )
sim.wt <- weighted_time( sim.rw, function (x) { x>0 } )
# distribution of time above zero depending on starting point
plot(xinit,sim.wt,xlab="initial position", ylab="time above zero")
```
We can also get the predicted clines by solving the PDE.
Note that here we are approximating $p(x) = \min(1,\exp(-4x\sqrt{s}))$.
```{r pushed_pde,cached=TRUE}
thegrid <- setup.grid.1D(x.up=-15, x.down=15, N = 5000)
clinefn <- setup.prop.1D(grid=thegrid, func=function(x,s){ifelse(x>0,exp(-4*x*sqrt(s)),1.0)}, s=s)
hetfn <- setup.prop.1D(grid=thegrid, func=function(x){ifelse(x>0,1,0)})
yinit <- rep(1.0,thegrid$N)
tt <- seq(0,tau,length.out=100)
pde.solns <- lapply( rvals, function (r) {
        pde.fn <- function (t,y,parms,...) {
            tran <- tran.1D(C=y, A=clinefn, D=1/2, flux.up=0, flux.down=0, dx=thegrid)$dC
            if (any(is.na(tran))) { browser() }
            list( tran - r * hetfn$mid * y )
        }
        ode.1D( y=yinit, times=tt, func=pde.fn, nspec=1 ) # note FIRST COLUMN IS TIME
      } )
layout(t(1:2))
for (k in c(6,10)) {
    clineplot(pde.solns[[k]],main=sprintf("r=%0.2f",tau,rvals[k]))
}
```

Here are the two compared,
by smoothing the probabilities for the simulation:
```{r sim_clines, cache=TRUE, dependson=c("sim_pushed","pushed_pde")}
prob.no.recomb <- sapply( rvals, function (r) { fitted(loess( exp(-r*sim.wt) ~ xinit, span=0.1)) } )
matplot(xinit,prob.no.recomb,type='l',ylab="prob of no recomb",
        xlab="space",col=rainbow(length(rvals)),lty=1,main=sprintf("t=%d",tau))
legend("bottomleft",lty=1:5,col=1:6,legend=paste("r=",round(rvals,digits=2)))
for (k in seq_along(rvals)) {
    lines( thegrid$x.mid, pde.solns[[k]][findInterval(tau,pde.solns[[k]][,1]),-1], 
          lty=3, lwd=3, col=rainbow(length(rvals))[k] )
}
```
Well, that looks reasonable, but hard to tell if the discrepancy is due to the loess or not.


Analytical form
---------------

If we define
$$\begin{aligned}
\Gamma_t = \text{meas}\{ t : X_t > 0 \} ,
\end{aligned}$$
then we want to know
$$\begin{aligned}
    u(x;\lambda,\theta):=\E^x \left[ \exp\left( - \lambda \Gamma_t \right) \right] .
\end{aligned}$$
(*Small problem:*  $X$ isn't guarenteed to be on the correct side of zero when we want it to be.)
Well, by Girsanov's theorem, we know that
$\E[f(X_t)] = \E[f(B_t)Z(t)]$
where
$$\begin{aligned}
Z(t) = \exp\left[ \int_0^t \theta(B_u) dB_t - \frac{1}{2} \int_0^t \theta(B_u)^2 du \right] 
\end{aligned}$$
and since by Tanaka's formula,
$$\begin{aligned}
\theta \left(\max(0,B_t) - \max(0,B_0)\right) 
&= 
\int_0^t \theta(B_u) du + \theta \ell_t(0) \\
&= 
\theta \Gamma_t + \theta \ell_t(0) ,
\end{aligned}$$
thus
$$\begin{aligned}
Z(t) = \exp\left[ \theta \max(0,B_t) - \theta \max(0,B_0) - \theta \ell_t(0) - \frac{1}{2} \int_0^t \theta(B_u)^2 du \right] 
\end{aligned}$$
and we want to know
$$\begin{aligned}
u(x;\lambda,\theta)
&=
\E^x \left[
\exp\left( 
- \lambda \Gamma_t
\right)
\right]
\end{aligned}$$

More than we need to know
-------------------------

According to Karatzas & Shreve (1984)
(with their $\delta=\theta_1=0$),
if $dX_t = -\theta(X_t) dt + dB_t$,
where $\theta(x)=\theta$ for $x \ge 0$, then
if we define
$$
\begin{aligned}
h(s;x,\mu) = \frac{|x|}{\sqrt{2\pi s^3}} \exp\left[-\frac{(x+\mu s)^2}{2s}\right] \qquad x>0 ,
\end{aligned}
$$
(this is the density of the first time a BM with drift $\mu$ started at $x$ hits 0),
then
$$
\begin{aligned}
p_t(x,z) 
&= \P\{ X_t = z \mid X_0 = x \}\\
&= \begin{cases}
    2 \int_0^\infty \int_0^t h(t-\tau;b-z,0) h(\tau;x+b,\theta) d\tau db
        & \qquad x \ge 0, \quad z\le 0 \\
    2 \int_0^\infty \int_0^t e^{2\theta z} h(t-\tau;b,0) h(\tau;x+b,\theta) d\tau db & \\
    \qquad {} + \frac{1}{\sqrt{2\pi t}} 
        \left\{ 
            \exp\left[ -\frac{ (x-z+\theta t)^2 }{2t} \right]
            +\exp\left[ -\frac{ (x+z-\theta t)^2 }{2t} - 2 \theta x \right]
        \right\}
        & \qquad x \ge 0, \quad z > 0 .
\end{cases}
\end{aligned}
$$
Also, $p_t(x,z;\theta)=p_t(-x,-z;-\theta)$ gets the same expressions for $x<0$.


Notes on excursion theory
=========================

Let $n(e)$ denote the excursion measure of Brownian motion,
on $\calE = \{ f : [0,\infty) \to [0,\infty) \, \text{s.t.} f(s)=0, t>s \Rightarrow f(t)=0 \}$.
For $f \in \calE$ denote $\zeta(f) = \sup\{ t : f(t)>0\}$.
Then [Rogers 1989](http://www.bibsonomy.org/bibtex/201686d46785a10af443e4d48a5196088/peter.ralph)
$$\begin{aligned}
n(f : \sup_t f(t) > a ) = \frac{1}{2a} .
\end{aligned}$$
and
$$\begin{aligned}
n(f : \zeta f > t ) = \sqrt{\frac{2}{\pi t}} .
\end{aligned}$$


