---
title: "Clines from PDE"
date: "`r date()`"
---

```{r doc_setup, include=FALSE}
fig.dim <- 5
knitr::opts_chunk$set(fig.width=2*fig.dim,fig.height=fig.dim,fig.align='center')
library(ReacTran)
library(Matrix)
source("tran.1D.R") # fix for log.A
```

*Useful functions used below:*
```{r setup_fns}
source("cline-fns.R")
```

The goal
========


Suppose that a lineage moves in a frequency profile $p(x)$ of allele $A$,
switching between backgrounds at rate $r p(x)$ or $r (1-p(x))$.
Let $(X_t,Z_t)$ denote this process,
so that
$$\begin{aligned}
    dX_t &= \grad \log(p(X_t)) dt + dB_t  \qquad \text{if } Z_t = A \\
         &= \grad \log(1-p(X_t)) dt + dB_t  \qquad \text{if } Z_t = B \\
    \P\{ Z_{t+\epsilon} = B | Z_t = A \} &= \epsilon r (1-p(X_t)) + O(\epsilon^2) \\
    \P\{ Z_{t+\epsilon} = A | Z_t = B \} &= \epsilon r p(X_t) + O(\epsilon^2) .
\end{aligned}$$
(Note that $p$ may also depend on $t$.)
We would like to find the probability that 
a lineage which we sample at location $x$ on background $z$,
derived from background $A$ $\tau$ units of time ago,
which we denote by
$$
q(x,z,\tau,r) = \P^x\{ Z_\tau = A \} .
$$
Also note that by symmetry, we should have that
if $r>0$, then
$$\begin{aligned}
    q(x,z,\tau,r) \to \frac{1}{2} \qquad \text{as } \tau \to \infty .
\end{aligned}$$

Let $f$ be a function from $\R \times \{A,B\} \to \R$.
The process above has generator
$$\begin{aligned}
    Gf(x,z) 
    &:= \partial_t \E^{(x,z)}\left[ f(X_t,Z_t) \right] \vert_{t=0} \\
    &= \begin{cases}
    \grad \log(p(x)) \cdot \grad f(x,A) + \frac{\sigma^2}{2} \Delta f(x,A) + r (1-p(x)) (f(x,B)-f(x,A)) \qquad &\text{if } z=A \\
    \grad \log(1-p(x)) \cdot \grad f(x,B) + \frac{\sigma^2}{2} \Delta f(x,B) + r p(x) (f(x,A)-f(x,B)) \qquad &\text{if } z=B .
    \end{cases}
\end{aligned}$$

We know that 
$$\begin{aligned}
q(x,A,0,r) &= 1  \\
q(x,B,0,r) &= 0  \\
\end{aligned}$$
and that $\partial_t q = G q$, i.e.,
(writing $q(x,A,t,r) = q_A(x,t)$ for the moment):
$$\begin{aligned}
    \partial_t q_A(x,t)
        &= 
        \grad \log(p(x)) \cdot \grad q_A(x,t) + \frac{\sigma^2}{2} \Delta q_A(x,t) + r (1-p(x)) (q_B(x,t)-q_A(x,t)) \\
    \partial_t q_B(x,t)
        &= 
        \grad \log(1-p(x)) \cdot \grad q_B(x,t) + \frac{\sigma^2}{2} \Delta q_B(x,t) + r p(x) (q_A(x,t)-q_B(x,t)) .
\end{aligned}$$

The cline shape
---------------

The PDE above gives probabilities given the initial background,
so therefore *linkage* to the selected site.
For the cline shape we need to average over choice of background,
which we denote
$$\begin{aligned}
p(x,\tau,r) =  p(x,\tau) q(x,A,\tau,r) + (1-p(x,\tau))q(x,B,\tau,r) .
\end{aligned}$$
Note that
$$\begin{aligned}
q(x,A,\tau,r) &= 1-q(-x,B,\tau,r) \\
p(x,\tau) &= 1-p(-x,\tau) 
\end{aligned}$$
and so
$$\begin{aligned}
p(x,\tau,r) =  p(x,\tau) q(x,A,\tau,r) + p(-x,\tau) (1-q(-x,A,\tau,r)) .
\end{aligned}$$




Paving the way for analytic approximations
------------------------------------------

Note that $p$ appears twice here:
first pushing lineages, and second causing recombination.
We may want to decouple these for the purposes of approximation:
we'll consider the following system of PDE:
$$\begin{aligned}
    \partial_t q_A(x,t)
        &= 
        \grad \log(u(x)) \cdot \grad q_A(x,t) + \frac{\sigma^2}{2} \Delta q_A(x,t) + r (1-v(x)) (q_B(x,t)-q_A(x,t)) \\
    \partial_t q_B(x,t)
        &= 
        \grad \log(1-u(x)) \cdot \grad q_B(x,t) + \frac{\sigma^2}{2} \Delta q_B(x,t) + r v(x) (q_A(x,t)-q_B(x,t)) \\
    q_A(x,0) &= 1 \\
    q_A(x,t) &= 1-q_B(-x,t) .
\end{aligned}$$

Here's a function to solve this numerically:
```{r solve_pde_setup, cache=TRUE}
thegrid <- setup.grid.1D(x.up=-100, x.down=100, N = 5000)
log_u <- function(x,...){ tanh_cline(x=-x,...,log=TRUE) }
log_1mu <- function(x,...){ tanh_cline(x=x,...,log=TRUE) }
cline_from_soln
solve_pde
```


Numerical solutions
-------------------

Let's solve this numerically.
```{r solve_pde, cache=TRUE, dependson="solve_pde_setup"}
r <- 0.1
s <- 0.1
sigma <- 5
zone.age <- 200  # this is tau
tt <- seq(0,zone.age,length.out=100)
pde.soln <- solve_pde( u=log_u, um1=log_1mu, r=r, times=tt, 
                  s=s, sigma=sigma, grid=thegrid, log.u=TRUE )
image(pde.soln,legend=TRUE,main=sprintf("q for z=%s, r=%0.3f",c("A","B"),r), yaxt='n', ylab='space',zlim=c(0,1))
axis(2,at=(pretty(thegrid$x.mid)-min(thegrid$x.mid))/diff(range(thegrid$x.mid)),labels=pretty(thegrid$x.mid))
```
These combine to give the predicted cline:
```{r show_pde_cline,cache=TRUE,dependson=c("solve_pde_setup","solve_pde")}
pde.cline <- cline_from_soln(pde.soln,grid=thegrid)
image(pde.cline,legend=TRUE,zlim=c(0,1),main=sprintf("clines, at r=%0.2f",r),yaxt='n')
axis(2,at=(pretty(thegrid$x.mid)-min(thegrid$x.mid))/diff(range(thegrid$x.mid)),labels=pretty(thegrid$x.mid))
# same thing, plotted with lines
clineplot(pde.cline,x=thegrid$x.mid,main=sprintf("clines, at r=%0.2f, up to t=%d",r,zone.age))
```
**Note:** the range should be wide enough these are *going to zero*
at the ends, or else we have to worry about the boundary conditions.

The smaller $r$ is, the slower the per-background probabilities smooth out:
```{r small_r_sim,cache=TRUE}
r <- 0.01
pde.soln <- solve_pde( u=log_u, um1=log_1mu, r=r, times=tt, 
                  s=s, sigma=sigma, grid=thegrid, log.u=TRUE )
image(pde.soln,legend=TRUE,main=sprintf("q for z=%s, r=%0.3f",c("A","B"),r), yaxt='n', ylab='space',zlim=c(0,1))
axis(2,at=(pretty(thegrid$x.mid)-min(thegrid$x.mid))/diff(range(thegrid$x.mid)),labels=pretty(thegrid$x.mid))
pde.cline <- cline_from_soln(pde.soln,grid=thegrid)
layout(1)
clineplot(pde.cline,x=thegrid$x.mid,main=sprintf("clines, at r=%0.2f, up to t=%d",r,zone.age))
```

And, the cline flattens out *very* slowly:
```{r longer_t_sim,cache=TRUE}
r <- 1.0
zone.age <- 2000  # this is tau
tt <- seq(0,zone.age,length.out=100)
pde.soln <- solve_pde( u=log_u, um1=log_1mu, r=r, times=tt, 
                  s=s, sigma=sigma, grid=thegrid, log.u=TRUE )
image(pde.soln,legend=TRUE,main=sprintf("q for z=%s, r=%0.3f",c("A","B"),r), yaxt='n', ylab='space',zlim=c(0,1))
axis(2,at=(pretty(thegrid$x.mid)-min(thegrid$x.mid))/diff(range(thegrid$x.mid)),labels=pretty(thegrid$x.mid))
pde.cline <- cline_from_soln(pde.soln,grid=thegrid)
layout(1)
clineplot(pde.cline,x=thegrid$x.mid,main=sprintf("clines, at r=%0.2f, up to t=%d",r,zone.age))
```


Haplotype lengths
=================

Consider a genomic region $[a,b]$,
with the selected site at 0.
Recombination without coalescence makes the history of this region
a branching process with conserved branching rates.
Each branch is labeled by a region of the genome,
and moves independently according to generator $G_r$, 
where $r$ is the distance from the region to the selected site,
and branches at rate $w$, where $w$ is its width.
We would like to know the probability that 
*all* branches end up on the same side of the barrier at time $\tau$ in the past.

Suppose that the region includes a type $A$ allele,
and we want the probability that it inherits entirely from the $A$ side.
Let
$$\begin{aligned}
    g(x,A,a,b,\tau) = \P^{x,A}\{ [a,b] \text{ inherits from $A$, $\tau$ generations ago }\} .
\end{aligned}$$
If $W(t)=\{W_1(t),W_2(t),\ldots,W_{n(t)}(t)\}$ is the state of the process $t$ generations ago,
where $W_k=(x_k,z_k,a_k,b_k)$,
with $x_k$ the position and $W_k\in\{A,B\}$ the allele,
then by independence of branches,
$$\begin{aligned}
    g(x,A,a,b,0) &= 1 \\
    g(x,B,a,b,0) &= 0 \\
    g(x,A,a,b,\tau)
    &=
    \E\left[
        \prod_{i=1}^{n(t)} g(x_k,z_k,a_k,b_k,\tau-t)
    \right] \qquad \text{for each } t \le \tau .
\end{aligned}$$
Taking $t = \epsilon$ in the last equation,
so that the probability of branching is $1-e^{-(b-a)\epsilon}=\epsilon(b-a)+O(\epsilon^2)$,
if $0\le a < b$ this implies that
$$\begin{aligned}
\frac{d}{d\tau} 
g(x,A,a,b,\tau)
    &=
    G_{r(a,b)} g(x,A,a,b,\tau)
    +
     \int_a^b g(x,A,a,\theta,\tau)
     \left(
     p(x,\tau)g(x,A,\theta,b,\tau) 
     + (1-p(x,\tau))g(x,B,\theta,b,\tau) 
     \right)d\theta
     -(b-a)g(x,A,a,b,\tau)
\end{aligned}$$
where 
$$\begin{aligned}
r(a,b)=\begin{cases}
    \min(|a|,|b|) \qquad &\text{if } ab \ge 0 \\
    0 \qquad &\text{if } 0 \in [a,b] .
\end{cases}
\end{aligned}$$
and $G_r f(x,z)$ is defined above.
On the other hand, if $a<0<b$, then
$$\begin{aligned}
\frac{d}{d\tau} 
g(x,A,a,b,\tau)
    &=
    G_{r(a,b)} g(x,A,a,b,\tau)
    +
     \int_0^b g(x,A,a,\theta,\tau)
     \left(
     p(x,\tau)g(x,A,\theta,b,\tau)
     + (1-p(x,\tau))g(x,B,\theta,b,\tau) 
     \right)d\theta \\
   &\qquad {}
    +
     \int_a^0 \left(
     p(x,\tau)g(x,A,a,\theta,\tau)
     + (1-p(x,\tau))g(x,B,a,\theta,\tau) 
     \right) g(x,A,\theta,b,\tau) d\theta  \\
   &\qquad {}
     -(b-a)g(x,A,a,b,\tau)
\end{aligned}$$

Writing $g_A$ and $g_B$ for the moment, and dropping $x$ and $\tau$,
the generator of the recombination part is, 
where $x\wedge y=\max(x,y)$ and $x \vee y = \min(x,y)$,
$$\begin{aligned}
Ig_A(a,b)
    &= 
        p \int_{a\wedge 0}^b g_A(a,\theta) g_A(\theta,b) d\theta
        + (1-p) \int_{a\wedge 0}^b g_A(a,\theta) g_B(\theta,b) d\theta
        + p \int_{a}^{b \vee 0} g_A(a,\theta) g_A(\theta,b) d\theta
        + (1-p) \int_{a\wedge 0}^b g_B(a,\theta) g_A(\theta,b) d\theta 
       \\ &\qquad {}
         -(b-a)g_A(a,b) \\
    &= 
    \int_a^b
        p \one_{\theta>0} g_A(a,\theta) g_A(\theta,b) 
        + (1-p) \one_{\theta>0} g_A(a,\theta) g_B(\theta,b) 
        + p \one_{\theta\le0} g_A(a,\theta) g_A(\theta,b) 
        + (1-p) \one_{\theta\le0} g_B(a,\theta) g_A(\theta,b) d\theta
       \\ &\qquad {}
         -(b-a)g_A(a,b) \\
    &= 
    \int_a^b
    \left\{
        p g_A(a,\theta) g_A(\theta,b) 
        + (1-p) \left( 
            \one_{\theta>0} g_A(a,\theta) g_B(\theta,b) 
            + \one_{\theta\le0} g_B(a,\theta) g_A(\theta,b) 
        \right) 
        -g_A(a,b) \right\}
        d\theta   
\end{aligned}$$
$Ig_B$ is the same after exchanging $A$ with $B$, and $p$ with $1-p$.

Therefore, assuming $g$ is sufficiently nice,
$$\begin{aligned}
\frac{d}{d\tau} g(x,z,a,b,\tau)
&=
G_{r(a,b)} g(x,z,a,b,\tau)
+
I g(x,z,a,b,\tau),
\end{aligned}$$
where $I$ is the integral operator defined above.
Since $z \in \{A,B\}$, this is a system of two integro-differential equations 
in three variables plus time ($x$, $a$, $b$, and $\tau$; four if $x\in\R^2$).

*Note:* We also know that
$$\begin{aligned}
    g(x,B,a,b,\tau) &= 0 \qquad \text{if } 0 \in [a,b] \\
\end{aligned}$$
but this will have to be arranged by the cline function.

Numerical solutions will discretize; what discretization of the genome might be good?
If we want to solve out to time $\tau$,
then on a chromosome of length 1 there will be about $\tau$ breakpoints,
each of which approximately Exponential($\tau$),
and so the smallest distance will have mean $1/\tau^2$.
But, suppose we've solved it for time $\tau$, and want to solve it to time $2\tau$.
We can use the same discretization, with initial conditions from the first round.
<!--
Quick check:
```{r smallest_check}
f <- function (t) { min(diff(c(0,sort(runif(rpois(1,lambda=t))),1))) }
fk <- sapply( 2:200, function (k) { replicate(n=100,f(k)) } )
plot(1+col(fk),fk,col=rainbow(300)[col(fk)],log='y',ylab='smallest interval')
points(2:200,colMeans(fk),pch=20,cex=2)
lines(2:200,1/(2:200)^2,col='purple',lwd=2)
```
-->

Note that if we think of $u(a,b)$ as a kernel, 
then the map $u(a,b) \mapsto \int_a^b u(a,\theta) u(\theta,b) d\theta$
is similar to the product of the kernel to itself;
on discretization this becomes like a matrix power.
This is implemented in some C code.
```{r solve_branching, cache=TRUE}
recomb_generator

s <- 0.1
tt <- seq(0,10,length.out=20)

xgrid <- setup.grid.1D(x.up=-20, x.down=20, N=10)
rgrid <- setup.grid.1D(x.up=-0.5, x.down=0.5, N=40)

# took 24min with xgrid$N=10, rgrid$N=20 and 50 time steps up to t=3, :(
hap.soln <- forwards_backwards_haplotypes(s=s, times=tt, xgrid=xgrid, rgrid=rgrid)
```


Here are haplotype probabilities centered on the selected site:
```{r plot_hap_probs, cache=TRUE, depends="solve_branching"}
for (haplen in c(0.02, 0.26, 0.46)) {
    hap.probs <- get_hap_probs( hap.soln, left=-haplen/2, right=haplen/2 )
    image( hap.probs, zlim=c(0,1), legend=TRUE, yaxt='n', 
          main=paste(c("A:","B:"),paste(round(attr(hap.probs,"r"),3),collapse=" to ")) )
    axis(2,at=(pretty(xgrid$x.mid)-min(xgrid$x.mid))/diff(range(xgrid$x.mid)),labels=pretty(xgrid$x.mid))
}
```

And, here are haplotype probabilities for a fairly short haplotype moving away from the selected site:
```{r plot_hap_probs_locs, cache=TRUE, depends="solve_branching"}
haplen <- 0.04
for (haploc in c(0.01, 0.09, 0.21, 0.31)) {
    hap.probs <- get_hap_probs( hap.soln, left=haploc, right=haploc+haplen )
    image( hap.probs, zlim=c(0,1), legend=TRUE, yaxt='n', 
          main=paste(c("A:","B:"),paste(round(attr(hap.probs,"r"),3),collapse=" to ")) )
    axis(2,at=(pretty(xgrid$x.mid)-min(xgrid$x.mid))/diff(range(xgrid$x.mid)),labels=pretty(xgrid$x.mid))
    locator(1)
}
```

Here's all the solution traces, i.e., how probabilities of inheriting an $A$ haplotype
of all lengths, positions, and spatial locations changes with time
```{r plot_hap_everything, cache=TRUE, depends="solve_branching"}
matplot(hap.soln[,1],hap.soln[,-1],type='l',xlab='time',ylab='probability')
```

Here's how these look across space, at a fixed time,
with line type showing haplotype length
and color showing distance to the selected site.
```{r plot_hap_clines, cache=TRUE, depends="solve_branching"}
plot.t <- 3
hap.clines <- hap.soln[which(hap.soln[,1]>=plot.t)[1],-1]
dim(hap.clines) <- attr(hap.soln,"soln.dims")
r.df <- data.frame(attr(hap.soln,"r"))
r.df$length <- r.df$right-r.df$left
r.df$dist <- ifelse(r.df$right*r.df$left>0,pmin(abs(r.df$right),abs(r.df$left)),0)
r.df$col <- as.numeric(cut(r.df$dist,6))
r.df$lty <- as.numeric(cut(r.df$length,5))
matplot(xgrid$x.mid,hap.clines[1:xgrid$N,],type='l',main=paste("A: t=",plot.t),ylim=c(0,1),xlab="space",col=r.df$col,lty=r.df$lty,ylab="probability")
legend("bottomleft",lty=unique(r.df$lty),legend=paste("length=",round(tapply(r.df$length,r.df$lty,mean),2)))
matplot(xgrid$x.mid,hap.clines[xgrid$N+1:xgrid$N,],type='l',main=paste("B: t=",plot.t),ylim=c(0,1),xlab="space",col=r.df$col,lty=r.df$lty,ylab="probability")
legend("topright",lty=1,col=unique(r.df$col),legend=paste("dist=",round(tapply(r.df$dist,r.df$col,mean),2)))
```

And, here's a heatmap of haplotype length across space and time:
shown is the longest haplotype length that has probability of at least 50% of coming entirely from $A$:
```{r plot_hap_lens, cache=TRUE, depends=c("solve_branching","plot_hap_clines")}
xfac <- rep(1:xgrid$N,nrow(r.df))
ABfac <- rep( rep(c("A","B"),each=xgrid$N), nrow(r.df) )
hap.lens <- lapply( c(A="A",B="B"), function (allele) {
    do.call( rbind, lapply( 1:nrow(hap.soln), function (k) {
            tapply( hap.soln[k,1+which(ABfac==allele)], xfac, function (xx) {
                   if(any(xx>0.5)) { max(r.df$length[xx>0.5]) } else { 0.0 }
                } )
        } ) )
    } )
filled.contour(hap.soln[,1],xgrid$x.mid,hap.lens$A,xlab="time",ylab="space",zlim=c(0,max(hap.lens$B)),main="A-linked A haplotype length")
filled.contour(hap.soln[,1],xgrid$x.mid,hap.lens$B,xlab="time",ylab="space",zlim=c(0,max(hap.lens$B)),main="B-linked A haplotype length")
```


Linkage
=======

The way the discretization for haplotypes is implemented
effectively only keeps track of a finite number of markers.
To do two-site ancestral LD, then,
we just need to use the above code at only two sites on the genome.

First, we compute per-allele solutions:
```{r solve_LD, cache=TRUE}
s <- 0.1
tt <- seq(0,100,length.out=21)
xgrid <- setup.grid.1D(x.up=-20.5, x.down=20.5, N=41)
rr <- seq(0,0.5,length.out=20)[-1]
fwds.soln <- forwards_pde(s=s,times=tt,grid=extend_grid(xgrid))
LD.solns <- lapply( rr, function (r) forwards_backwards_haplotypes(s=s, times=tt, xgrid=xgrid, rgrid=list(x.mid=c(0,r),N=2), fwds.soln=fwds.soln) )
```
... but then we must extract linkage from these.
Below we report for recombination distance $r$,
the probability that two markers at distance $r$ 
sampled in one individual at that location and time
(with one end fixed to the selected site)
are both of the same ancestral type.
If the local proportion of type $A$ is $p$ and the probability when linked to $A$ is $u_A$,
we want $p u_A(r_0,r_0+r) + (1-p) u_B(r_0,r_0+r) - (p u_A(r_0) + (1-p) u_A(r_0))(p u_A(r_0+r) + (1-p) u_A(r_0+r))$:
```{r compute_LD, cache=TRUE, depends="solve_LD"}
pfun <- function (t,x) { 
        fp <- function (x,t) { approx(xgrid$x.mid,cline_interp(t,soln=fwds.soln),xout=x)$y }
        ans <- numeric(length(x))
        for (tval in unique(t)) {
            dothese <- (rep_len(t,length(x))==tval)
            ans[dothese] <- fp(x=x[dothese],t=tval)
        }
        return(ans)
}
ABfac <- rep( rep(c("A","B"),each=xgrid$N), 3 )  # three combinations of two recomb points
rfac <- rep( c("11","12","22"), each=xgrid$N*2 )
LDs <- lapply( LD.solns, function (lds) {
            # returns a time x space matrix
            pp <- outer( lds[,1], xgrid$x.mid, pfun )
            return( 
                   pp * lds[,1+which(ABfac=="A"&rfac=="12")] + (1-pp) * lds[,1+which(ABfac=="B"&rfac=="12")] 
                   - ( pp * lds[,1+which(ABfac=="A"&rfac=="11")] + (1-pp) * lds[,1+which(ABfac=="B"&rfac=="11")] )
                       * ( pp * lds[,1+which(ABfac=="A"&rfac=="22")] + (1-pp) * lds[,1+which(ABfac=="B"&rfac=="22")] )
               )
    } )
plot.x <- xgrid$x.mid[which.min(abs(xgrid$x.mid))]
matplot( tt, sapply( LDs, function (x) { x[,match(plot.x,xgrid$x.mid)] } ), type='l', ylab="ancestral LD", xlab="time", main=paste("location=",round(plot.x,2)) )
plot.t <- tt[which.min(abs(tt-20))]
matplot( xgrid$x.mid, sapply( LDs, function (x) { x[match(plot.t,tt),] } ), type='l', ylab="ancestral LD", xlab="space" )
matplot( rr, t(sapply( LDs, function (x) { x[match(plot.t,tt),] } )), type='l', ylab="ancestral LD", xlab="recomb dist", main=paste("t=",round(plot.t,2)), col=rainbow(xgrid$N), lty=1 )
legend("topright",lty=1,col=rainbow(xgrid$N)[floor(seq(1,xgrid$N,length.out=6))],legend=paste("x=",round(xgrid$x.mid[floor(seq(1,xgrid$N,length.out=6))],2)))
```
Hm, the right/left spatial positions should look the same, right?  Is something off?

**Note:** do the same with fixed distance between the two and changing distance to selected locus.


Distance to closest recombination in a heterozygote
===================================================

A simpler statistic, that might approximate haplotype statistics well,
is the distance along the chromosome from an $A$ allele until the closest recombination
that occurred in an $AB$ heterozygote.
Let $H$ be this distance;
then, 
$$\begin{aligned}
    u(x,\tau,r)=\P^{(x,\tau)}\{H>r\}
\end{aligned}$$
solves the following PDE:
$$\begin{aligned}
    u(x,0,r) &= 1  \\
    \partial_t u(x,t,r)
        &= 
        \grad \log(p(x)) \cdot \grad u(x,t,r) + \frac{\sigma^2}{2} \Delta u(x,t,r) - r (1-p(x)) u(x,t,r) .
\end{aligned}$$




Comparison to simulation
========================

We'll setup, run, and summarize simulations
as described in [this file](../sims/run-sims.html).
```{r setup_sims, cache=TRUE}
source("../sims/sim-fns.R")
zone.age    = 50
ndemes 		= 500
deme.size 	= 5
ninds 		= ndemes*deme.size
qtl         = list(chr1=data.frame(traits=c("underdominant"), s = s, pos=c(0.5)))
# spatial locations
xx <- (1:ndemes)-0.5-ndemes/2
# run the simulation
sim <- sim.zone(n.gen=zone.age,n.ind=ninds,n.deme=ndemes,sigma=sigma,quiet=TRUE)
# post-process
sim.sums <- spBreaks(sim)
```
and, postprocess:
```{r genotype_sims, cache=TRUE, dependson="setup_sims"}
rvals <- c(0,exp(seq(log(.001),log(0.5),length.out=10)))
loci <- list(0.5+c(rev(rvals[-1]),-rvals))
genotypes <- data.frame(do.call(cbind,lapply(sim.sums$ind.ancest, geno.ind, loci)))
freqs <- apply(genotypes,1,function(Z){
            tapply(Z,cut(1:ncol(genotypes),breaks=seq(0,ncol(genotypes),deme.size)),mean)/2
        })
```

OK, now let's match the PDE solution to the simulation.
First, does the cline at the selected locus match the the tanh cline?
```{r match_cline}
plot(xx,freqs[,1],type='l',xlab="spatial location",ylab="proportion")
lines(xx,1-tanh_cline(xx,s=s,sigma=sigma),col='red')
```
Looks pretty reasonable.

Here's the PDE solution:
```{r match_pde,cache=TRUE,dependson="solve_pde_setup"}
zone.age <- sim$pars["n.gen"]
tt <- seq(0,zone.age,length.out=100)
pde.solns <- lapply(rvals, function (r) { solve_pde( u=log_u, um1=log_1mu, r=r, times=tt, 
                                                s=s, sigma=sigma, grid=thegrid, log.u=TRUE ) } )
pde.clines <- lapply( pde.solns, cline_from_soln, grid=thegrid )
match.clines <- sapply(pde.clines,function(x)x[length(tt),-1])
for (k in floor(seq(2,length(rvals),length.out=3))) {
    image(pde.solns[[k]],legend=TRUE,main=sprintf("r=%f",rvals[k]))
}
```
Now, comparing the PDE solution to the others:
```{r match_cline_pde}
for (k in seq_along(rvals)) {
    matplot(xx,freqs[,which(abs(rvals[k]-loci[[1]]+0.5)<1e-9)],type='l',
            xlab="spatial location",ylab="proportion",main=sprintf("r=%0.3f",rvals[k]))
    lines(xx,1-tanh_cline(xx,s=s,sigma=sigma),col='red',lwd=3,lty=3)
    lines(1-thegrid$x.mid,match.clines[,k],col='green',lwd=3)
}
```
