---
title: "Rough calculations"
date: "`r date()`"
---

$$
\newcommand{\E}{\mathbb{E}}
\renewcommand{\P}{\mathbb{P}}
\newcommand{\R}{\mathbb{R}}
\DeclareMathOperator{\sgn}{sgn}
\newcommand{\grad}{\nabla}
$$

```{r doc_setup, include=FALSE}
fig.dim <- 5
knitr::opts_chunk$set(fig.width=2*fig.dim,fig.height=fig.dim,fig.align='center')
library(ReacTran)
source("tran.1D.R") # fix for log.A
```

In the following, $x$ is the distance from the cline center,
$r$ is the recombination rate to the selected locus,
$\tau$ is the time since secondary contact.
The two alleles are $A$ and $B$.
We would like to know:

1.  **How a neutral cline flattens:**  
    What is the width of the cline at a linked, neutral locus?
    How long does it take to flatten out?
    How far from the cline do you have to go to find mostly one type?

2.  **How do haplotypes look around the selected locus in the cline:**
    How long are the haplotypes around the rare $B$ alleles on the $A$ side?
    This should be constant in time.

3.  **How does introgression happen:**
    How long, and how common, are chunks of $B$ haplotype *not* linked to $B$ alleles on the $A$ side?
    These will accumulate as time progresses.

4.  **What can we learn from linkage?**
    For a pair of sites, what is the correlation in ancestral state at those loci,
    given the state at the selected locus
    (and, also, averaged over choice of selected locus).

    a.  Allow for ascertainment of SNPs in extremal individuals
        by estimating probability of coalescence in ancestral populations
        from the extremal individuals.
        For each pair of loci, correlation of these two loci with extremal individuals.

5.  How does neighborhood size affect the results?
    Does the mean cline, averaged across low-population-density simulations,
    match the cline obtained by a high population density, i.e., the PDE?

6.  What about the variance??



How a neutral cline flattens
============================


1.  *What is the width of the cline at a linked, neutral locus?*
    *How long does it take to flatten out?* *How far from the cline do you have to go to find mostly one type?*

**Answer:**

a.  Maximum cline width $\sigma \sqrt{\tau}$, even at unlinked loci.
b.  The selected cline takes time $1/s$ to spread out to its full width.
c.  Cline behaves like the selected cline for $r < 1/\tau$, and begin to flatten for $r>1/\tau$.
d.  For large distances and times (but with $x<\sigma\sqrt{\tau}$),
    the cline may stay steep out to
    $r = \sqrt{\frac{2s}{\tau}}$ rather than $r=1/\tau$,
    but this seems hard to actually observe.

**Rationale:**

a.  The maximum of Brownian motion until time $t$ has mean $\sqrt{2t/\pi}$,
    and has the same marginal distribution as the absolute value of Brownian motion.
    Therefore, the probability a lineage from $x$ hasn't got to the center of the cline *at all* by time $\tau$ 
    is about $2\Phi(x/\sqrt{\sigma^2 \tau})-1$;
    and therefore, if $x/\sigma > 2\sqrt{\tau}$ then the cline cannot have started to form, regardless of $r$.

b.  The selected cline spreads out (in forwards time!) with diffusion,
    and so will reach maximum width $\sigma/\sqrt{s}$ in time $1/s$.
    Let $q(t,x)$ be the solution to
    $$\begin{aligned}
    \dot q = \frac{1}{2} q'' + q (1-q) (2q-1) .
    \end{aligned}$$
    If $u=at$ and $y=bx$ and let $p(t,x) = q(at,bx)$ then
    $$\begin{aligned}
    \dot p(t,x)
        &= \partial_t q(at,bx) \\
        &= a \dot q(at,bx) \\
        &= a \left\{ \frac{1}{2} q''(at,bx) + q (1-q) (2q-1) \right\} \\
        &= a \left\{ \frac{1}{2b^2} \partial^2_x q(at,bx) + q (1-q) (2q-1) \right\} \\
        &= \frac{a}{2b^2} p''(t,x) + a p (1-p) (2p-1)  .
    \end{aligned}$$
    If we choose $a=s$ and $b=\sigma/\sqrt{s}$ then $a/b^2=\sigma^2$, so
    $$\begin{aligned}
    \dot p = \frac{\sigma^2}{2} p'' + s p (1-p) (2p-1) .
    \end{aligned}$$
    We know that the unitless cline $q$ will spread out in time 1;
    so our cline will spread out in time $1/s$.

c.  The cline will be narrower if $r$ is small.
    Consider a lineage begun near the center of the cline.
    The most time it can spend in heterozygotes is $\tau$,
    so if $r \ll 1/\tau$, the chance of recombining is small.
    So, the cline will look like the cline at the selected locus if $r < 1/\tau$.

d.  By LÃ©vy's identity, the local time of a Brownian motion at zero
    has the same distribution (pathwise!) as the running maximum of Brownian motion.
    Therefore, the local time at zero up until $\tau$ grows like $\sqrt{\tau}$.
    The same is true for Brownian motion reflected at zero.
    The time a lineage spends in heterozygotes should be roughly proportional to the local time
    (each time it visits zero from the left, it spends some time in hets before leaving again; 
    the constant of proportionality is the mean time spent per visit).
    Therefore, there will be appreciable chance of recombination if $r C \sqrt{\tau}>1$,
    where $C$ is the constant;
    this says that if clines down to $r$ are flat at $\tau$, then at time $2\tau$ they'll be flat down to $r/4$.

    But, what is the constant $C$?
    Consider the approximation of a lineage that acts like Brownian motion with a constant drift $s$ above zero.
    As noted in [this file](diffusion-local-times.html),
    the excursion measure of Brownian motion assigns mass $\sqrt{\frac{1}{2 \pi t^3}}$ to excursions of length $t$;
    we must tilt this by $e^{-ts}$ for the selection above zero;
    so that the mean rate at which this semi-pushed motion accumulates time above zero,
    with respect to local time at zero,
    is  (letting $x=\sqrt{2st}$)
    $$\begin{aligned}
    \int_0^\infty t \frac{e^{-st}}{ \sqrt{2 \pi t^3 } } dt 
    &=
    \int_0^\infty \frac{e^{-st}}{ \sqrt{2 \pi t } } dt \\
    &=
    \sqrt{2s} \int_0^\infty \frac{e^{-x^2/2}}{ \sqrt{2 \pi} } dx \\
    &=
    \frac{1}{2\sqrt{2s}} .
    \end{aligned}$$
    This suggests taking $C=\sqrt{s/2}$.


Simulation check
----------------

We'll check these with the PDE solutions
(and, then with simulations):

```{r setup_fns}
source("cline-fns.R")
```



a.  Look at space-time frequency images at $r=1$ and check they spread out as $\sqrt{t}$.
    The green lines are at multiples of $\pm \sqrt{t}$.
    
    ```{r unlinked_clines, cache=TRUE}
    thegrid <- setup.grid.1D(x.up=-50, x.down=50, N = 5000)
    s <- 0.01
    r <- 1.0
    tt <- seq(0,1000,length.out=250)
    unlinked.solns <- forwards_backwards_pde(s=s,times=tt,grid=thegrid,r=r,log.p=TRUE)
    # image(unlinked.solns,legend=TRUE,zlim=c(0,1),yaxt='n')
    # axis(2,at=(pretty(thegrid$x.mid)-min(thegrid$x.mid))/diff(range(thegrid$x.mid)),labels=pretty(thegrid$x.mid))
    unlinked.cline  <- cline_from_soln( unlinked.solns, grid=thegrid )
    image(unlinked.cline,legend=TRUE,zlim=c(0,1),yaxt='n',main=sprintf("r=%f, s=%f",r,s))
    axis(2,at=(pretty(thegrid$x.mid)-min(thegrid$x.mid))/diff(range(thegrid$x.mid)),labels=pretty(thegrid$x.mid))
    matlines(tt,0.5+outer(sqrt(tt)/diff(range(thegrid$x.mid)),(1:4)/2,"*"),lwd=2,col='green')
    matlines(tt,0.5-outer(sqrt(tt)/diff(range(thegrid$x.mid)),(1:4)/2,"*"),lwd=2,col='green')
    ```

b.  Look at space-time frequency images at $r=0$ and several values of $s$
    to check they become stationary after about $1/s$.
    The clines are colored red for times less than $1/s$, and black afterwards;
    the green dotted lines are at $\pm 1/\sqrt{s}$.
    Indeed, cline flattening has slowed considerably by $1/s$, 
    and the limiting cline reaches $(1/4,3/4)$ at $\pm 1/\sqrt{s}$.
    
    ```{r forwards_selected_cline,cache=TRUE}
    thegrid <- setup.grid.1D(x.up=-100, x.down=100, N = 5000)
    svals <- exp(seq(log(1e-4),log(0.5),length.out=6))
    fwds.solns <- lapply(svals, function(s) {
            tt <- seq(0,3/s,length.out=40)
            forwards_pde(s=s,times=tt,grid=thegrid) 
        } )
    for (k in seq_along(svals)) {
        clineplot( fwds.solns[[k]], main=sprintf("s=%f",svals[k]), col=ifelse( fwds.solns[[k]][,1]>1/svals[k], adjustcolor("black",0.5), adjustcolor("red",0.5) ) )
        abline(v=c(-1,1)/sqrt(svals[k]),lty=3,col='green',lwd=2)
    }
    ```

c.  To check that flattening occurs around $r=1/\tau$ (i.e., at $\tau=1/r$)
    we'll look at when a neutral allele linked to the $A$ allele
    near the center starts to get close to 1/2.
    
    (*Technical note:* the system of ODE is [stiff](https://en.wikipedia.org/wiki/Stiff_equation), 
    as diffusive systems [usually are](http://scicomp.stackexchange.com/questions/891/the-definition-of-stiff-ode-system),
    particularly for small $r$ at the times when each set of probabilities becomes steep;
    setting $s=0.1$ this is a problem, but not with smaller (less steep) $s$.)
    
    ```{r space_recomb, cache=TRUE}
    s <- 0.01
    rvals <- c(0,exp(seq(log(.0025),log(0.05),length.out=8)))
    tt <- seq(0,1000,length.out=100)
    thegrid <- setup.grid.1D(x.up=-50, x.down=50, N = 2000)
    fwds.soln <- forwards_pde(s=s,times=seq(0,1.2*max(tt),length.out=1000),grid=thegrid)
    r.solns <- lapply(rvals, function(r) {
            forwards_backwards_pde(s=s,times=tt,grid=thegrid,r=r,log.p=TRUE,fwds.soln=fwds.soln)
        } )
    linked.clines  <- lapply(r.solns, cline_from_soln, grid=thegrid )
    xr <- lapply( seq_along(tt), function (k) { do.call( cbind, lapply( linked.clines, function (x) x[k,-1] ) ) } )
    ```
    The vertical lines are at $1/r$ (only on the righthand plots, due to the limitations of `filled.contour`):
    ```{r plot_sr}
    for (k in seq_along(rvals)) {
        image( r.solns[[k]], legend=TRUE, zlim=c(0,1), yaxt='n', ylab='position', main=sprintf("r=%f, s=%f",rvals[k],s=s) )
        axis(2,at=(pretty(thegrid$x.mid)-min(thegrid$x.mid))/diff(range(thegrid$x.mid)),labels=pretty(thegrid$x.mid))
        abline(v=1/rvals[k],lwd=2,lty=3)
    }
    ```

d.  To look for the $r=1/\sqrt{\tau}$ behaviour, we could look at positions outside the selected cline 
    ($x$ larger than $C/\sqrt{2s}$, for C=4, say) 
    and times large enough that unlinked clines have started to flatten out there 
    and that the selected cline has got to full width
    ($\tau$ larger than $C^2/s$).
    But: if $\tau=C^2/s$ then the critical value is $r=(s/C)^2$,
    which at $s=0.01$ is $r=6.25\times10^{-6}$
    **Moral:** the narrower cline due to linkage may be only visible only very nearby to the selected site.
    *To-do:* figure out parameter values where this behaviour becomes visible: need to be much further away?

<!--
    ```{r long_time,cache=TRUE}
    s <- 0.01
    r <- 0.0001
    tt <- seq(0,30000,length.out=500)
    thegrid <- setup.grid.1D(x.up=-200, x.down=200, N = 2000)
    fwds.soln <- forwards_pde(s=s,times=seq(0,1.2*max(tt),length.out=1000),grid=thegrid)
    rev.soln <- forwards_backwards_pde(s=s,times=tt,grid=thegrid,r=r,log.p=TRUE,fwds.soln=fwds.soln)
    linked.cline  <- cline_from_soln( rev.soln, grid=thegrid )
    ```
-->



How do haplotypes look around the selected locus in the cline
=============================================================

2.  How long are the haplotypes around the rare $B$ alleles on the $A$ side?

**Answer:**

a.  The haplotype of a $B$ allele at distance $x$ into the $A$ side should be of length about $\sigma/(sx)$.


**Rationale:**

a.  This is (probably) the Laplace transform of the time since the $B$ allele
    last had an ancestor on the $B$ side,
    i.e., for a $B$ lineage to make it back to the $B$ side.
    Since lineages on the wrong side move at speed $\sigma/s$,
    if the allele is at position $x$, 
    this will take about $sx/\sigma$;
    and so the haplotype should be about $\sigma/(sx)$.
